<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Changelog • PhenotypeSimulator</title>

<link rel="shortcut icon" type="image/x-icon" href="../favicon.ico">

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.1/clipboard.min.js" integrity="sha256-hIvIxeqhGZF+VVeM55k0mJvWpQ6gTkWk3Emc+NmowYA=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Changelog" />



<!-- mathjax -->
<script src='https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-news">
      <header>
      <div class="navbar navbar-standard navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">PhenotypeSimulator</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-file"></span>
     
    documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/PhenotypeSimulator.html">Step-by-step phenotype simulation</a>
    </li>
    <li>
      <a href="../articles/Simulation-and-LinearModel.html">Application of PhenotypeSimulator for a power analysis study</a>
    </li>
    <li>
      <a href="../articles/Sample-scripts-external-genotype-simulation.html">Genotype simulation supported by PhenotypeSimulator</a>
    </li>
    <li>
      <a href="../articles/SimulationBasedOnExampleData.html">Phenotype simulation based on an example dataset</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="../news/index.html">
    <span class="fa fa-newspaper-o"></span>
     
    news
  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-code"></span>
     
    functions
  </a>
</li>
<li>
  <a href="https://github.com/HannahVMeyer/PhenotypeSimulator">
    <span class="fa fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1>Changelog <small></small></h1>
      <small>Source: <a href='https://github.com/HannahVMeyer/PhenotypeSimulator/blob/master/NEWS.md'><code>NEWS.md</code></a></small>
    </div>

    <div id="phenotypesimulator-0-3-0" class="section level1">
<h1 class="page-header">
<a href="#phenotypesimulator-0-3-0" class="anchor"></a>PhenotypeSimulator 0.3.0<small> Unreleased </small>
</h1>
<div id="major-changes" class="section level2">
<h2 class="hasAnchor">
<a href="#major-changes" class="anchor"></a>Major changes</h2>
<ol>
<li>Add option for non-linear transformation of simulated phenotypes: function (transformNonlinear)[<a href="https://github.com/HannahVMeyer/PhenotypeSimulator/blob/master/R/createphenotypeFunctions.R" class="uri">https://github.com/HannahVMeyer/PhenotypeSimulator/blob/master/R/createphenotypeFunctions.R</a>], accessible from <a href="https://github.com/HannahVMeyer/PhenotypeSimulator/blob/master/R/createphenotypeFunctions.R">runSimulation</a>. Both transformed and original phenotypes are automatically returned with <a href="https://github.com/HannahVMeyer/PhenotypeSimulator/blob/master/R/outputFunctions.R">savePheno</a>
</li>
<li>
<p>Replace parameter ‘oxgen’ in <a href="https://github.com/HannahVMeyer/PhenotypeSimulator/blob/master/R/genotypeFunctions.R">readStandardGenotypes</a> and <a href="https://github.com/HannahVMeyer/PhenotypeSimulator/blob/master/R/genotypeFunctions.R">getCausalSNPs</a> with ‘format’ - ensures proper specification of genotype format for all cases.</p>
<h2 id="minor-changes">Minor changes</h2>
</li>
<li>In addition to full kinship, <a href="https://github.com/HannahVMeyer/PhenotypeSimulator/blob/master/R/outputFunctions.R">savePheno</a> and <a href="https://github.com/HannahVMeyer/PhenotypeSimulator/blob/master/R/outputFunctions.R">writeStandardOutput</a> write eigenvalues and eigenvalues of kinship matrix.</li>
<li>Output file names have been made more consistent in <a href="https://github.com/HannahVMeyer/PhenotypeSimulator/blob/master/R/outputFunctions.R">savePheno</a> and <a href="https://github.com/HannahVMeyer/PhenotypeSimulator/blob/master/R/outputFunctions.R">writeStandardOutput</a>.</li>
<li>Causal SNPs are now also saved in specified standard output format.</li>
<li><p>LiMMBo has been added as output format in <a href="https://github.com/HannahVMeyer/PhenotypeSimulator/blob/master/R/outputFunctions.R">savePheno</a> and <a href="https://github.com/HannahVMeyer/PhenotypeSimulator/blob/master/R/outputFunctions.R">writeStandardOutput</a>(<a href="https://limmbo.readthedocs.io/en/latest/">LiMMBo format</a>)</p></li>
</ol>
</div>
</div>
    <div id="phenotypesimulator-0-2-2" class="section level1">
<h1 class="page-header">
<a href="#phenotypesimulator-0-2-2" class="anchor"></a>PhenotypeSimulator 0.2.2<small> 2018-05-13 </small>
</h1>
<div id="minor-changes-1" class="section level2">
<h2 class="hasAnchor">
<a href="#minor-changes-1" class="anchor"></a>Minor changes</h2>
<ol>
<li>Update readStandardGenotypes to be compatible with latest release of data.table (v1.11.2), see <a href="https://github.com/HannahVMeyer/PhenotypeSimulator/issues/7">here</a>
</li>
</ol>
</div>
</div>
    <div id="phenotypesimulator-0-2-1" class="section level1">
<h1 class="page-header">
<a href="#phenotypesimulator-0-2-1" class="anchor"></a>PhenotypeSimulator 0.2.1<small> Unreleased </small>
</h1>
<div id="minor-changes-2" class="section level2">
<h2 class="hasAnchor">
<a href="#minor-changes-2" class="anchor"></a>Minor changes</h2>
<ol>
<li>Additional tests for compatibility of input parameters with variance components functions, genotype functions and output functions.</li>
<li>Bug fix in output function: savePheno now properly saves kinship matrix as .rds.</li>
</ol>
</div>
</div>
    <div id="phenotypesimulator-0-2-0" class="section level1">
<h1 class="page-header">
<a href="#phenotypesimulator-0-2-0" class="anchor"></a>PhenotypeSimulator 0.2.0<small> 2018-01-07 </small>
</h1>
<div id="major-changes-1" class="section level2">
<h2 class="hasAnchor">
<a href="#major-changes-1" class="anchor"></a>Major changes</h2>
<p><strong>Input</strong></p>
<ol>
<li>
<em>PhenotypeSimulator</em> now includes readStandardGenotypes which can read externally simulated or user-provided genotypes in plink, genome, oxgen (hapgen/impute2), bimbam or simple delimited format.</li>
<li>A user-specified correlation matrix can be provided for the simulation of the correlatedBdEffects.</li>
<li>Short option flags for command-line use of <em>PhenotypeSimulator</em> were removed.</li>
</ol>
<p><strong>Output</strong></p>
<ol>
<li>
<em>PhenotypeSimulator</em> provides the option to save the simulated phenotypes and genotypes in formats compatible with a number of commonly used genetic association software (gemma, bimbam, plink, snptest) via writeStandardOutput.</li>
<li>Intermediate phenotype components are now saved per default.</li>
<li>Saving additional subsets of the simulated data has been removed.</li>
</ol>
<p><strong>Variance components</strong></p>
<ol>
<li><p>Genotype simulation and kinship estimation: functions for genotype simulation and kinship estimation have been rewritten for significant speed-ups of the computation time <a href="https://github.com/HannahVMeyer/PhenotypeSimulator-profiling">benchmarking</a>.</p></li>
<li>geneticFixedEffects and noiseFixedEffects:
<ol>
<li><p>The effect size distributions of the shared effects are now modelled as the product of two exponential distributions (to yield an approximately uniform distributions) or the product of a normal distribution with user-specified parameters and a standard normal distribution.</p></li>
<li><p>The independent effects can now be specified to affect the same subset or different subsets of traits (via keepSameIndependent).</p></li>
<li><p>The overall number of traits affected by the effects can now be specified via pTraitsAffected.</p></li>
</ol>
</li>
<li><p>correlatedBgEffects: the additional correlation between the traits can be specified by the user by providing an external correlation matrix.</p></li>
</ol>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
        <li><a href="#phenotypesimulator-0-3-0">0.3.0</a></li>
        <li><a href="#phenotypesimulator-0-2-2">0.2.2</a></li>
        <li><a href="#phenotypesimulator-0-2-1">0.2.1</a></li>
        <li><a href="#phenotypesimulator-0-2-0">0.2.0</a></li>
      </ul>
    </div>
  </div>

</div>

      <footer>
      <div class="copyright">
  <p>Developed by Hannah Meyer.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

