<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Flexible phenotype simulation with PhenotypeSimulator • PhenotypeSimulator</title>
<link rel="shortcut icon" type="image/x-icon" href="../favicon.ico">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.1/clipboard.min.js" integrity="sha256-hIvIxeqhGZF+VVeM55k0mJvWpQ6gTkWk3Emc+NmowYA=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Flexible phenotype simulation with PhenotypeSimulator">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-standard navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">PhenotypeSimulator</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-file"></span>
     
    documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/PhenotypeSimulator.html">Step-by-step phenotype simulation</a>
    </li>
    <li>
      <a href="../articles/Simulation-and-LinearModel.html">Application of PhenotypeSimulator for a power analysis study</a>
    </li>
    <li>
      <a href="../articles/Sample-scripts-external-genotype-simulation.html">Genotype simulation supported by PhenotypeSimulator</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="../news/index.html">
    <span class="fa fa-newspaper-o"></span>
     
    news
  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-code"></span>
     
    functions
  </a>
</li>
<li>
  <a href="https://github.com/HannahVMeyer/PhenotypeSimulator">
    <span class="fa fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Flexible phenotype simulation with PhenotypeSimulator</h1>
                        <h4 class="author">Hannah Meyer</h4>
            
            <h4 class="date">2018-10-24</h4>
      
      
      <div class="hidden name"><code>PhenotypeSimulator.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>PhenotypeSimulator allows for the flexible simulation of phenotypes from genetic and noise components. In quantitative genetics, genotype to phenotype mapping is commonly realised by fitting a linear model to the genotype as the explanatory variable and the phenotype as the response variable. Other explanatory variables such as additional sample measures (e.g. age, height, weight) or batch effects can also be included. For linear mixed models, in addition to the fixed effects of the genotype and the covariates, different random effect components can be included, accounting for population structure in the study cohort or observational noise. The application of linear and linear mixed models in quantitative genetics ranges from genetic studies in model organisms such as yeast and <em>Arabidopsis thaliana</em> to human molecular, morphological or imaging-derived traits. Developing new methods to efficiently model increasing sample sizes or to apply multi-variate models to sets of phenotypic measurements often requires simulated datasets with a specific underlying phenotype structure.</p>
<p><em>PhenotypeSimulator</em> allows for the simulation of such phenotypes under different models, including genetic variant effects, infinitesimal genetic effects (reflecting population structure and kinship), non-genetic covariate effects, correlated noise effects and observational noise effects. Different phenotypic effects can be combined into a final phenotype while controling for the proportion of variance explained by each of the components. For each component, the number of variables, their distribution and the design of their effect across traits can be customised. The <strong>work-flow</strong> outlined below summarizes the strategy for the phenotype simulation. In section <strong>Examples</strong>, phenotype simulation for phenotypes with different levels of complexity are demonstrated in both a step-by-step manner or by using the recommended <code>runSimulation</code> function. Finally, section <strong>Phenotype component functions</strong> explains the use and simulation strategy of the individual phenotype component-generating functions.</p>
</div>
<div id="work-flow" class="section level1">
<h1 class="hasAnchor">
<a href="#work-flow" class="anchor"></a>Work-flow</h1>
<ol style="list-style-type: decimal">
<li>Simulate phenotype components of interest:
<ol style="list-style-type: decimal">
<li>
<code>geneticFixedEffects</code>: genetic variant effects,</li>
<li>
<code>geneticBgEffects</code>: population structure and kinship,</li>
<li>
<code>noiseFixedEffects</code>: confounding variables e.g sex, age, height…,</li>
<li>
<code>correlatedBgEffects</code>: additional correlation effects,</li>
<li>
<code>noiseBgEffects</code>: observational noise;</li>
</ol>
</li>
<li>scale components according to variance explained: each phenotype component is scaled to explain a certain proportion of the entire phenotypic variance via <code>rescaleVariance</code>;</li>
<li>combine phenotype components: rescaled phenotype components are combined to obtain the final simulated phenotype.</li>
</ol>
<p><code>runSimulation</code> combines the three steps outlined above and allows for the automatic simulation of a phenotype with <span class="math inline">\(N\)</span> number of samples, <span class="math inline">\(P\)</span> number of traits and up to five phenotype components. Alternatively, all components can be simulated, scaled and addedd independently. <code>savePheno</code> accepts the output of <code>runSimulation</code> to save phenotype components, the final phenotype and -optionally- simulated genotypes in the specified directories.</p>
<p>The following sections outline two examples for phenotype simulations with population structure and observational noise effects and with a more complex set-up of the phenotypes from five phenotype components.</p>
<p>As demonstrated below, each phenotype component function has a number of parameters that allow for customisation of the simulation. <code>runSimulation</code> wraps around all these functions and accepts a multitude of parameters. Simple simulation of phenotypes, however, only requires the input of the desired phenotype size (number of samples and traits) and the proportion of variance each phenotype component should take. The recommended use of PhenotypeSimulator is via <code>runSimulation</code> as this ensures that all dependencies are automatically set correctly.</p>
<p>The functions used in these examples are explained in detail in <strong>Phenotype component functions</strong>.</p>
</div>
<div id="examples" class="section level1">
<h1 class="hasAnchor">
<a href="#examples" class="anchor"></a>Examples</h1>
<div id="example-1-creating-a-phenotype-composed-of-population-structure-and-observational-noise-effects-only-" class="section level2">
<h2 class="hasAnchor">
<a href="#example-1-creating-a-phenotype-composed-of-population-structure-and-observational-noise-effects-only-" class="anchor"></a>Example 1: Creating a phenotype composed of population structure and observational noise effects only.</h2>
<div id="step-by-step" class="section level3">
<h3 class="hasAnchor">
<a href="#step-by-step" class="anchor"></a>Step by step</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Set parameters</span>
genVar &lt;-<span class="st"> </span><span class="fl">0.4</span>
noiseVar &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">-</span><span class="st"> </span>genVar
shared &lt;-<span class="st"> </span><span class="fl">0.6</span>
independent &lt;-<span class="st"> </span><span class="dv">1</span> <span class="op">-</span><span class="st"> </span>shared

<span class="co"># simulate simple bi-allelic genotypes and estimate kinship</span>
genotypes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/simulateGenotypes.html">simulateGenotypes</a></span>(<span class="dt">N =</span> <span class="dv">100</span>, <span class="dt">NrSNP =</span> <span class="dv">10000</span>, 
                               <span class="dt">frequencies =</span> <span class="kw">c</span>(<span class="fl">0.05</span>, <span class="fl">0.1</span>, <span class="fl">0.3</span>, <span class="fl">0.4</span>), 
                               <span class="dt">verbose =</span> <span class="ot">FALSE</span>)
genotypes_sd &lt;-<span class="st"> </span><span class="kw"><a href="../reference/standardiseGenotypes.html">standardiseGenotypes</a></span>(genotypes<span class="op">$</span>genotypes)
kinship &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getKinship.html">getKinship</a></span>(<span class="dt">N=</span><span class="dv">100</span>, <span class="dt">X=</span>genotypes_sd, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)

<span class="co"># simulate phenotype components</span>
genBg &lt;-<span class="st"> </span><span class="kw"><a href="../reference/geneticBgEffects.html">geneticBgEffects</a></span>(<span class="dt">N =</span> <span class="dv">100</span>, <span class="dt">kinship =</span> kinship, <span class="dt">P =</span> <span class="dv">15</span>)
noiseBg &lt;-<span class="st"> </span><span class="kw"><a href="../reference/noiseBgEffects.html">noiseBgEffects</a></span>(<span class="dt">N =</span> <span class="dv">100</span>, <span class="dt">P =</span> <span class="dv">15</span>)

<span class="co"># rescale phenotype components</span>
genBg_shared_scaled &lt;-<span class="st"> </span><span class="kw"><a href="../reference/rescaleVariance.html">rescaleVariance</a></span>(genBg<span class="op">$</span>shared, shared <span class="op">*</span><span class="st"> </span>genVar)
genBg_independent_scaled &lt;-<span class="st"> </span><span class="kw"><a href="../reference/rescaleVariance.html">rescaleVariance</a></span>(genBg<span class="op">$</span>independent, 
                                            independent <span class="op">*</span><span class="st"> </span>genVar)
noiseBg_shared_scaled &lt;-<span class="st"> </span><span class="kw"><a href="../reference/rescaleVariance.html">rescaleVariance</a></span>(noiseBg<span class="op">$</span>shared, shared <span class="op">*</span><span class="st"> </span>noiseVar)
noiseBg_independent_scaled &lt;-<span class="st"> </span><span class="kw"><a href="../reference/rescaleVariance.html">rescaleVariance</a></span>(noiseBg<span class="op">$</span>independent, 
                                              independent <span class="op">*</span>noiseVar)

<span class="co"># Total variance proportion shave to add up yo 1</span>
total &lt;-<span class="st"> </span>independent <span class="op">*</span><span class="st"> </span>noiseVar <span class="op">+</span><span class="st"> </span>independent <span class="op">*</span><span class="st"> </span>genVar <span class="op">+</span><span class="st"> </span>
<span class="st">    </span>shared <span class="op">*</span><span class="st"> </span>noiseVar <span class="op">+</span><span class="st"> </span>shared <span class="op">*</span><span class="st"> </span>genVar

total <span class="op">==</span><span class="st"> </span><span class="dv">1</span>
<span class="co">#&gt; [1] TRUE</span>

<span class="co"># combine components into final phenotype</span>
Y &lt;-<span class="st"> </span><span class="kw">scale</span>(genBg_shared_scaled<span class="op">$</span>component <span class="op">+</span><span class="st"> </span>noiseBg_shared_scaled<span class="op">$</span>component <span class="op">+</span>
<span class="st">    </span>genBg_independent_scaled<span class="op">$</span>component <span class="op">+</span><span class="st"> </span>noiseBg_independent_scaled<span class="op">$</span>component)

<span class="co"># transform phenotype non-linearly</span>
Y_nl &lt;-<span class="st"> </span><span class="kw"><a href="../reference/transformNonlinear.html">transformNonlinear</a></span>(Y, <span class="dt">alpha=</span><span class="fl">0.7</span>, <span class="dt">method=</span><span class="st">"exp"</span>)</code></pre></div>
</div>
<div id="via-runsimulation" class="section level3">
<h3 class="hasAnchor">
<a href="#via-runsimulation" class="anchor"></a>via <code>runSimulation</code>
</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># simulate phenotype with population structure and observational noise effects </span>
<span class="co"># only</span>

<span class="co"># genetic variance</span>
genVar &lt;-<span class="st"> </span><span class="fl">0.4</span>

<span class="co"># random genetic variance: h2b </span>
phenotype &lt;-<span class="st"> </span><span class="kw"><a href="../reference/runSimulation.html">runSimulation</a></span>(<span class="dt">N =</span> <span class="dv">100</span>, <span class="dt">P =</span> <span class="dv">15</span>,  <span class="dt">tNrSNP =</span> <span class="dv">10000</span>, 
                           <span class="dt">SNPfrequencies =</span> <span class="kw">c</span>(<span class="fl">0.05</span>, <span class="fl">0.1</span>,<span class="fl">0.3</span>,<span class="fl">0.4</span>), 
                           <span class="dt">genVar =</span> <span class="fl">0.4</span>, <span class="dt">h2bg =</span> <span class="dv">1</span>, <span class="dt">phi =</span> <span class="dv">1</span>, 
                           <span class="dt">verbose =</span> <span class="ot">TRUE</span>, <span class="dt">nonlinear=</span><span class="st">"exp"</span>, 
                           <span class="dt">proportionNonlinear =</span> <span class="fl">0.7</span>)
<span class="co">#&gt; Set seed: 219453</span>
<span class="co">#&gt; The total noise variance (noiseVar) is: 0.6</span>
<span class="co">#&gt; The noise model is: noiseBgOnly</span>
<span class="co">#&gt; Proportion of observational noise variance (phi): 1</span>
<span class="co">#&gt; Variance of shared observational noise effect (alpha): 0.8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; The total genetic variance (genVar) is: 0.4</span>
<span class="co">#&gt; The genetic model is: geneticBgOnly</span>
<span class="co">#&gt; Proportion of variance of infinitesimal genetic effects (h2bg): 1</span>
<span class="co">#&gt; Proportion of variance of shared infinitesimal genetic effects (eta): 0.8</span>
<span class="co">#&gt; Proportion of non-linear phenotype transformation (proportionNonlinear): 0.7</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Simulate genetic effects (genetic model: geneticBgOnly)</span>
<span class="co">#&gt; Simulate 10000 SNPs...</span>
<span class="co">#&gt; Standardising the 10000 SNPs provided</span>
<span class="co">#&gt; Estimating kinship from 10000 SNPs provided</span>
<span class="co">#&gt; Normalising kinship</span>
<span class="co">#&gt; Simulate infinitesimal genetic effects</span>
<span class="co">#&gt; Simulate noise terms (noise model: noiseBgOnly)</span>
<span class="co">#&gt; Simulate observational noise effects</span>
<span class="co">#&gt; Construct final simulated phenotype</span></code></pre></div>
</div>
</div>
<div id="example-2-creating-a-phenotype-composed-of-five-components" class="section level2">
<h2 class="hasAnchor">
<a href="#example-2-creating-a-phenotype-composed-of-five-components" class="anchor"></a>Example 2: Creating a phenotype composed of five components</h2>
<p>The phenoytpes to be simulated are composed of genetic variant effects, infinitesimal genetic effects (reflecting population structure and kinship), non-genetic covariate effects, correlated noise effects and observational noise effects.</p>
<div id="step-by--step" class="section level3">
<h3 class="hasAnchor">
<a href="#step-by--step" class="anchor"></a>step-by- step</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># read genotypes from external file</span>
<span class="co"># use one of the sample genotype file provided in the </span>
<span class="co"># extdata/genotypes/subfolders (e.g.extdata/genotypes/hapgen )</span>
genotypefile &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata/genotypes/hapgen"</span>, 
                            <span class="st">"genotypes_hapgen.controls.gen"</span>, 
                            <span class="dt">package =</span> <span class="st">"PhenotypeSimulator"</span>)
<span class="co"># remove the .gen ending (oxgen specific endings .gen and .sample are added </span>
<span class="co"># automatically )</span>
genotypefile &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">"</span><span class="ch">\\</span><span class="st">.gen"</span>,<span class="st">""</span>, genotypefile)

genotypes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/readStandardGenotypes.html">readStandardGenotypes</a></span>(<span class="dt">N=</span><span class="dv">100</span>, <span class="dt">filename =</span> genotypefile,
                                    <span class="dt">format=</span><span class="st">"oxgen"</span>,
                                    <span class="dt">delimiter =</span> <span class="st">","</span>, <span class="dt">verbose=</span><span class="ot">TRUE</span>)

genotypes_sd &lt;-<span class="st"> </span><span class="kw"><a href="../reference/standardiseGenotypes.html">standardiseGenotypes</a></span>(genotypes<span class="op">$</span>genotypes)

<span class="co"># kinship estimate based on standardised SNPs </span>
kinship &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getKinship.html">getKinship</a></span>(<span class="dt">N=</span><span class="dv">100</span>, <span class="dt">X=</span>genotypes_sd, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)

<span class="co"># simulate 30 genetic variant effects (from non-standardised SNP genotypes)</span>
causalSNPs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getCausalSNPs.html">getCausalSNPs</a></span>(<span class="dt">N=</span><span class="dv">100</span>, <span class="dt">genotypes =</span> genotypes<span class="op">$</span>genotypes, 
                            <span class="dt">NrCausalSNPs =</span> <span class="dv">30</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)
genFixed &lt;-<span class="st"> </span><span class="kw"><a href="../reference/geneticFixedEffects.html">geneticFixedEffects</a></span>(<span class="dt">N =</span> <span class="dv">100</span>, <span class="dt">P =</span> <span class="dv">15</span>, <span class="dt">X_causal =</span> causalSNPs)  

<span class="co"># simulate infinitesimal genetic effects</span>
genBg &lt;-<span class="st"> </span><span class="kw"><a href="../reference/geneticBgEffects.html">geneticBgEffects</a></span>(<span class="dt">N=</span><span class="dv">100</span>, <span class="dt">kinship =</span> kinship, <span class="dt">P =</span> <span class="dv">15</span>)

<span class="co"># simulate 4 different confounder effects:</span>
<span class="co"># * 1 binomial covariate effect shared across all traits</span>
<span class="co"># * 2 categorical (3 categories) independent covariate effects</span>
<span class="co"># * 1 categorical (4 categories) independent  covariate effect</span>
<span class="co"># * 2 normally distributed independent and shared covariate effects</span>
noiseFixed &lt;-<span class="st"> </span><span class="kw"><a href="../reference/noiseFixedEffects.html">noiseFixedEffects</a></span>(<span class="dt">N =</span> <span class="dv">100</span>, <span class="dt">P =</span> <span class="dv">15</span>, <span class="dt">NrFixedEffects =</span> <span class="dv">4</span>, 
                                <span class="dt">NrConfounders =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">2</span>),
                                <span class="dt">pIndependentConfounders =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="fl">0.5</span>),  
                                <span class="dt">distConfounders =</span> <span class="kw">c</span>(<span class="st">"bin"</span>, <span class="st">"cat_norm"</span>, 
                                                    <span class="st">"cat_unif"</span>, <span class="st">"norm"</span>),
                                <span class="dt">probConfounders =</span> <span class="fl">0.2</span>, 
                                <span class="dt">catConfounders =</span> <span class="kw">c</span>(<span class="dv">3</span>, <span class="dv">4</span>))

<span class="co"># simulate correlated effects with max correlation of 0.8</span>
correlatedBg &lt;-<span class="st"> </span><span class="kw"><a href="../reference/correlatedBgEffects.html">correlatedBgEffects</a></span>(<span class="dt">N =</span> <span class="dv">100</span>, <span class="dt">P =</span> <span class="dv">15</span>, <span class="dt">pcorr =</span> <span class="fl">0.8</span>)

<span class="co"># simulate observational noise effects</span>
noiseBg &lt;-<span class="st"> </span><span class="kw"><a href="../reference/noiseBgEffects.html">noiseBgEffects</a></span>(<span class="dt">N =</span> <span class="dv">100</span>, <span class="dt">P =</span> <span class="dv">15</span>)

<span class="co"># total SNP effect on phenotype: 0.01</span>
genVar &lt;-<span class="st"> </span><span class="fl">0.6</span>
noiseVar &lt;-<span class="st"> </span><span class="dv">1</span> <span class="op">-</span><span class="st"> </span>genVar
totalSNPeffect &lt;-<span class="st"> </span><span class="fl">0.01</span>
h2s &lt;-<span class="st"> </span>totalSNPeffect<span class="op">/</span>genVar
phi &lt;-<span class="st"> </span><span class="fl">0.6</span> 
rho &lt;-<span class="st"> </span><span class="fl">0.1</span>
delta &lt;-<span class="st"> </span><span class="fl">0.3</span>
shared &lt;-<span class="st"> </span><span class="fl">0.8</span>
independent &lt;-<span class="st"> </span><span class="dv">1</span> <span class="op">-</span><span class="st"> </span>shared

<span class="co"># rescale phenotype components</span>
genFixed_shared_scaled &lt;-<span class="st"> </span><span class="kw"><a href="../reference/rescaleVariance.html">rescaleVariance</a></span>(genFixed<span class="op">$</span>shared, shared <span class="op">*</span><span class="st"> </span>h2s <span class="op">*</span>genVar)
genFixed_independent_scaled &lt;-<span class="st"> </span><span class="kw"><a href="../reference/rescaleVariance.html">rescaleVariance</a></span>(genFixed<span class="op">$</span>independent, 
                                            independent <span class="op">*</span><span class="st"> </span>h2s <span class="op">*</span>genVar)
genBg_shared_scaled &lt;-<span class="st"> </span><span class="kw"><a href="../reference/rescaleVariance.html">rescaleVariance</a></span>(genBg<span class="op">$</span>shared, shared <span class="op">*</span><span class="st"> </span>(<span class="dv">1</span><span class="op">-</span>h2s) <span class="op">*</span>genVar)
genBg_independent_scaled &lt;-<span class="st"> </span><span class="kw"><a href="../reference/rescaleVariance.html">rescaleVariance</a></span>(genBg<span class="op">$</span>independent, 
                                            independent <span class="op">*</span><span class="st"> </span>(<span class="dv">1</span><span class="op">-</span>h2s) <span class="op">*</span><span class="st"> </span>genVar)

noiseBg_shared_scaled &lt;-<span class="st"> </span><span class="kw"><a href="../reference/rescaleVariance.html">rescaleVariance</a></span>(noiseBg<span class="op">$</span>shared, shared <span class="op">*</span><span class="st"> </span>phi<span class="op">*</span><span class="st"> </span>noiseVar)
noiseBg_independent_scaled &lt;-<span class="st"> </span><span class="kw"><a href="../reference/rescaleVariance.html">rescaleVariance</a></span>(noiseBg<span class="op">$</span>independent, 
                                              independent <span class="op">*</span><span class="st"> </span>phi<span class="op">*</span><span class="st"> </span>noiseVar)
correlatedBg_scaled &lt;-<span class="st"> </span><span class="kw"><a href="../reference/rescaleVariance.html">rescaleVariance</a></span>(correlatedBg<span class="op">$</span>correlatedBg, 
                                       shared <span class="op">*</span><span class="st"> </span>rho <span class="op">*</span><span class="st"> </span>noiseVar)

noiseFixed_shared_scaled &lt;-<span class="st"> </span><span class="kw"><a href="../reference/rescaleVariance.html">rescaleVariance</a></span>(noiseFixed<span class="op">$</span>shared, shared <span class="op">*</span><span class="st"> </span>delta <span class="op">*</span><span class="st"> </span>
<span class="st">                                                </span>noiseVar)
noiseFixed_independent_scaled &lt;-<span class="st"> </span><span class="kw"><a href="../reference/rescaleVariance.html">rescaleVariance</a></span>(noiseFixed<span class="op">$</span>independent, 
                                              independent <span class="op">*</span><span class="st"> </span>delta <span class="op">*</span><span class="st"> </span>noiseVar)

<span class="co"># Total variance proportions have to add up yo 1</span>
total &lt;-<span class="st"> </span>shared <span class="op">*</span><span class="st"> </span>h2s <span class="op">*</span>genVar <span class="op">+</span><span class="st">  </span>independent <span class="op">*</span><span class="st"> </span>h2s <span class="op">*</span><span class="st"> </span>genVar <span class="op">+</span>
<span class="st">    </span>shared <span class="op">*</span><span class="st"> </span>(<span class="dv">1</span><span class="op">-</span>h2s) <span class="op">*</span><span class="st"> </span>genVar <span class="op">+</span><span class="st">   </span>independent <span class="op">*</span><span class="st"> </span>(<span class="dv">1</span><span class="op">-</span>h2s) <span class="op">*</span><span class="st"> </span>genVar <span class="op">+</span>
<span class="st">    </span>shared <span class="op">*</span><span class="st"> </span>phi<span class="op">*</span><span class="st"> </span>noiseVar <span class="op">+</span><span class="st">  </span>independent <span class="op">*</span><span class="st"> </span>phi<span class="op">*</span><span class="st"> </span>noiseVar <span class="op">+</span>
<span class="st">    </span>rho <span class="op">*</span><span class="st"> </span>noiseVar <span class="op">+</span>
<span class="st">    </span>shared <span class="op">*</span><span class="st"> </span>delta <span class="op">*</span><span class="st"> </span>noiseVar <span class="op">+</span><span class="st">  </span>independent <span class="op">*</span><span class="st"> </span>delta <span class="op">*</span><span class="st"> </span>noiseVar

total <span class="op">==</span><span class="st"> </span><span class="dv">1</span>
<span class="co">#&gt; [1] TRUE</span>

<span class="co"># combine components into final phenotype</span>
Y &lt;-<span class="st"> </span><span class="kw">scale</span>(genBg_shared_scaled<span class="op">$</span>component <span class="op">+</span><span class="st"> </span>noiseBg_shared_scaled<span class="op">$</span>component <span class="op">+</span>
<span class="st">        </span>genBg_independent_scaled<span class="op">$</span>component <span class="op">+</span><span class="st"> </span>noiseBg_independent_scaled<span class="op">$</span>component <span class="op">+</span>
<span class="st">        </span>genFixed_shared_scaled<span class="op">$</span>component <span class="op">+</span><span class="st"> </span>noiseFixed_shared_scaled<span class="op">$</span>component <span class="op">+</span>
<span class="st">        </span>genFixed_independent_scaled<span class="op">$</span>component <span class="op">+</span><span class="st"> </span>noiseFixed_independent_scaled<span class="op">$</span>component <span class="op">+</span>
<span class="st">        </span>correlatedBg_scaled<span class="op">$</span>component)</code></pre></div>
</div>
<div id="via-runsimulation-1" class="section level3">
<h3 class="hasAnchor">
<a href="#via-runsimulation-1" class="anchor"></a>via <code>runSimulation</code>
</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># simulate phenotype with the same five phenotype components</span>
<span class="co"># and settings as above; display progress via verbose=TRUE</span>
phenotype &lt;-<span class="st"> </span><span class="kw"><a href="../reference/runSimulation.html">runSimulation</a></span>(<span class="dt">N =</span> <span class="dv">100</span>, <span class="dt">P =</span> <span class="dv">15</span>, <span class="dt">genotypefile =</span> genotypefile, 
    <span class="dt">format =</span> <span class="st">"oxgen"</span>, <span class="dt">cNrSNP =</span> <span class="dv">30</span>, <span class="dt">genVar =</span> genVar, <span class="dt">h2s =</span> h2s, 
    <span class="dt">phi =</span> <span class="fl">0.6</span>, <span class="dt">delta =</span> <span class="fl">0.3</span>, <span class="dt">distBetaGenetic =</span> <span class="st">"unif"</span>, <span class="dt">mBetaGenetic =</span> <span class="fl">0.5</span>, 
    <span class="dt">sdBetaGenetic =</span> <span class="dv">1</span>, <span class="dt">NrFixedEffects =</span> <span class="dv">4</span>, <span class="dt">NrConfounders =</span> <span class="kw">c</span>(<span class="dv">1</span>, 
        <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">2</span>), <span class="dt">pIndependentConfounders =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="fl">0.5</span>), 
    <span class="dt">distConfounders =</span> <span class="kw">c</span>(<span class="st">"bin"</span>, <span class="st">"cat_norm"</span>, <span class="st">"cat_unif"</span>, <span class="st">"norm"</span>), 
    <span class="dt">probConfounders =</span> <span class="fl">0.2</span>, <span class="dt">catConfounders =</span> <span class="kw">c</span>(<span class="dv">3</span>, <span class="dv">4</span>), <span class="dt">pcorr =</span> <span class="fl">0.8</span>, 
    <span class="dt">verbose =</span> <span class="ot">TRUE</span>)
<span class="co">#&gt; Set seed: 219453</span>
<span class="co">#&gt; The total noise variance (noiseVar) is: 0.4</span>
<span class="co">#&gt; The noise model is: noiseFixedAndBgAndCorrelated</span>
<span class="co">#&gt; Proportion of non-genetic covariate variance (delta): 0.3</span>
<span class="co">#&gt; Proportion of variance of shared non-genetic covariate effects (gamma): 0.8</span>
<span class="co">#&gt; Proportion of non-genetic covariates to have a trait-independent effect (pIndependentConfounders ): 0 1 1 0.5</span>
<span class="co">#&gt; Proportion of traits influenced by independent non-genetic covariate effects (pTraitIndependentConfounders): 0.2</span>
<span class="co">#&gt; Proportion of variance of correlated noise effects (rho): 0.1</span>
<span class="co">#&gt; Proportion of observational noise variance (phi): 0.6</span>
<span class="co">#&gt; Variance of shared observational noise effect (alpha): 0.8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; The total genetic variance (genVar) is: 0.6</span>
<span class="co">#&gt; The genetic model is: geneticFixedAndBg</span>
<span class="co">#&gt; Proportion of variance of genetic variant effects (h2s): 0.0166666666666667</span>
<span class="co">#&gt; Proportion of variance of shared genetic variant effects (theta): 0.8</span>
<span class="co">#&gt; Proportion of genetic variant effects to have a trait-independent fixed effect (pIndependentGenetic): 0.4</span>
<span class="co">#&gt; Proportion of traits influenced by independent genetic variant effects (pTraitIndependentGenetic): 0.2</span>
<span class="co">#&gt; Proportion of variance of infinitesimal genetic effects (h2bg): 0.983333333333333</span>
<span class="co">#&gt; Proportion of variance of shared infinitesimal genetic effects (eta): 0.8</span>
<span class="co">#&gt; Proportion of non-linear phenotype transformation (proportionNonlinear): 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Simulate genetic effects (genetic model: geneticFixedAndBg)</span>
<span class="co">#&gt; Simulate genetic variant effects</span>
<span class="co">#&gt; Out of 15 total phenotypes, 15 traits will be affected by genetic variant effects</span>
<span class="co">#&gt; Out of these affected traits (15), 3 trait(s) will have independent genetic variant effects</span>
<span class="co">#&gt; Standardising the 1000 SNPs provided</span>
<span class="co">#&gt; Estimating kinship from 1000 SNPs provided</span>
<span class="co">#&gt; Normalising kinship</span>
<span class="co">#&gt; Simulate infinitesimal genetic effects</span>
<span class="co">#&gt; Simulate noise terms (noise model: noiseFixedAndBgAndCorrelated)</span>
<span class="co">#&gt; Simulate correlated background effects</span>
<span class="co">#&gt; Simulate observational noise effects</span>
<span class="co">#&gt; Simulate confounder effects</span>
<span class="co">#&gt; Out of 15 total phenotypes, 15 trait(s) will be affected by the  1  covariate effect</span>
<span class="co">#&gt; Out of 15 total phenotypes, 15 trait(s) will be affected by the  2  covariate effect</span>
<span class="co">#&gt; Out of these affected traits (15), 3 trait(s) will have independent covariate effects</span>
<span class="co">#&gt; Out of 15 total phenotypes, 15 trait(s) will be affected by the  3  covariate effect</span>
<span class="co">#&gt; Out of these affected traits (15), 3 trait(s) will have independent covariate effects</span>
<span class="co">#&gt; Out of 15 total phenotypes, 15 trait(s) will be affected by the  4  covariate effect</span>
<span class="co">#&gt; Out of these affected traits (15), 3 trait(s) will have independent covariate effects</span>
<span class="co">#&gt; Construct final simulated phenotype</span></code></pre></div>
<p>Proportion of variance of the different phenotype components in the final phenotype:</p>
<table class="table">
<caption>Proportion of variance explained by the different phenotype components</caption>
<thead><tr class="header">
<th></th>
<th align="right">shared effect</th>
<th align="right">independent effect</th>
<th align="right">total effect</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>varGenFixed</td>
<td align="right">0.008</td>
<td align="right">0.002</td>
<td align="right">0.01</td>
</tr>
<tr class="even">
<td>varGenBg</td>
<td align="right">0.472</td>
<td align="right">0.118</td>
<td align="right">0.59</td>
</tr>
<tr class="odd">
<td>varNoiseFixed</td>
<td align="right">0.096</td>
<td align="right">0.024</td>
<td align="right">0.12</td>
</tr>
<tr class="even">
<td>varNoiseBg</td>
<td align="right">0.192</td>
<td align="right">0.048</td>
<td align="right">0.24</td>
</tr>
<tr class="odd">
<td>varNoiseCorr</td>
<td align="right">0.040</td>
<td align="right">0.000</td>
<td align="right">0.04</td>
</tr>
<tr class="even">
<td>sumProportion</td>
<td align="right">0.808</td>
<td align="right">0.192</td>
<td align="right">1.00</td>
</tr>
</tbody>
</table>
<p>The heatmap images below show the values of the phenotype itself (left) and of the correlation between the phenotypic traits (right). The code to produce the images can be seen below (for all subsequent images of the same type the code is not echo-ed).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">### show 'image' of phenotype and correlation between phenotypic traits
<span class="kw">image</span>(<span class="kw">t</span>(phenotype<span class="op">$</span>phenoComponentsFinal<span class="op">$</span>Y), <span class="dt">main=</span><span class="st">"Phenotype: [samples x traits]"</span>, 
      <span class="dt">axes=</span><span class="ot">FALSE</span>, <span class="dt">cex.main=</span><span class="fl">0.8</span>)
<span class="kw">mtext</span>(<span class="dt">side =</span> <span class="dv">1</span>, <span class="dt">text =</span> <span class="st">"Samples"</span>, <span class="dt">line =</span> <span class="dv">1</span>)
<span class="kw">mtext</span>(<span class="dt">side =</span> <span class="dv">2</span>, <span class="dt">text =</span> <span class="st">"Traits"</span>, <span class="dt">line =</span> <span class="dv">1</span>)
<span class="kw">image</span>(<span class="kw">cor</span>(phenotype<span class="op">$</span>phenoComponentsFinal<span class="op">$</span>Y), 
      <span class="dt">main=</span><span class="st">"Correlation of traits [traits x traits]"</span>, <span class="dt">axes=</span><span class="ot">FALSE</span>, <span class="dt">cex.main=</span><span class="fl">0.8</span>)
<span class="kw">mtext</span>(<span class="dt">side =</span> <span class="dv">1</span>, <span class="dt">text =</span> <span class="st">"Traits"</span>, <span class="dt">line =</span> <span class="dv">1</span>)
<span class="kw">mtext</span>(<span class="dt">side =</span> <span class="dv">2</span>, <span class="dt">text =</span> <span class="st">"Traits"</span>, <span class="dt">line =</span> <span class="dv">1</span>)</code></pre></div>
<p><img src="PhenotypeSimulator_files/figure-html/unnamed-chunk-7-1.png" width="326.4"><img src="PhenotypeSimulator_files/figure-html/unnamed-chunk-7-2.png" width="326.4"></p>
<p>The final phenotype and all its components can be saved via <code>savePheno</code>. <code>savePheno</code> takes the output of <code>runSimulation</code> and saves its components, if saveIntermediate is set to TRUE, intermediate phenotype componentes (such as shared and inpendent variance components are saved); if genotypes were simulated or read from external files and a kinhsip estimated thereof, they will also be saved. The user needs to have writing permission to the specified genotype and phenotype directories. The code below saves the genotypes/phenotypes as .csv files into the subdirectory “test_simulation” of the directories /tmp/genotypes and /tmp/phenotypes. The genotypes are additionally saved in binary plink format, i.e. .bed, .bim and .fam. If the user has writing permissions and the directories do not exist yet, they will be created.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">out &lt;-<span class="st"> </span><span class="kw"><a href="../reference/savePheno.html">savePheno</a></span>(phenotype, <span class="dt">directory=</span><span class="st">"/tmp"</span>,
                 <span class="dt">outstring=</span><span class="st">"test_simulation"</span>,
                 <span class="dt">format=</span><span class="kw">c</span>(<span class="st">"csv"</span>, <span class="st">"plink"</span>), <span class="dt">verbose=</span><span class="ot">FALSE</span>)</code></pre></div>
</div>
</div>
</div>
<div id="command-line-use" class="section level1">
<h1 class="hasAnchor">
<a href="#command-line-use" class="anchor"></a>Command line use</h1>
<p>PhenotypeSimulator can also be run from the command line via</p>
<p><code>Rscript -e  "PhenotypeSimulator::simulatePhenotypes()" --args --...</code> with <code>--...</code> being the user supplied simulation paramaters. <code>Rscript -e  "PhenotypeSimulator::simulatePhenotypes()"</code> takes the same arguments as <code>runSimulation</code> and <code>savePheno</code>: first, it simulates the specified phenotype components and then saves them into to specified directories. The user will need to have writing permissions to these parent directores. If directoryGeno and directoryPheno do not exist yet, they will be created.</p>
<p><code>Rscript -e  "PhenotypeSimulator::simulatePhenotypes()" --args --help</code> will print information about possible input parameters and values they can take on. To generate the same phenotypes as described above via the command line-interface, run the following code from your command line:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">Rscript</span> -e <span class="st">"PhenotypeSimulator::simulatePhenotypes()"</span> \
--args \
--NrSamples=100 --NrPhenotypes=15 \
--tNrSNP=10000 --cNrSNP=30 \
--SNPfrequencies=0.05,0.1,0.3,0.4 \
--genVar=0.4 --h2s=0.025 --phi=0.6 --delta=0.3 --gamma=1 \
--pcorr=0.8 \
--NrFixedEffects=4 --NrConfounders=1,2,1,2 \
--pIndependentConfounders=0,1,1,0.5 \
--distConfounders=bin,cat_norm,cat_unif,norm \
--probConfounders=0.2 \
--catConfounders=3,4 \
--directory=/tmp \
--subdirectory=test_simulation \
--showProgress \
--saveTable \
--savePLINK</code></pre></div>
</div>
<div id="phenotype-component-functions" class="section level1">
<h1 class="hasAnchor">
<a href="#phenotype-component-functions" class="anchor"></a>Phenotype component functions</h1>
<div id="genetic-variant-effects" class="section level2">
<h2 class="hasAnchor">
<a href="#genetic-variant-effects" class="anchor"></a>1. Genetic variant effects:</h2>
<p>Genetic variant effects are simulated as the matrix product of an [N x NrCausalSNPs] genotype matrix and [NrSNP x P] effect size matrix. The genotype matrix can either be drawn from i) a simulated genotype matrix ii) genotypes read from external data or iii) causal SNPs can be randomly sampled as lines from existing genotype files. In the latter case, genotypes are expected to be stored in a [SNPs x N] format, with separate files for each chromosome. The user can either specify which chromosomes to sample the SNPs from or simply provide the total number of chromosomes to sample from. For the simulation of genotypes, the user can specify the <code>NrSNP</code>s to simulate and a vector of allele frequencies <code>frequencies</code>. These allele frequencies are uniformly sampled and bi-allelic SNPs are simulated, with the sampled allele frequency acting as the probability in a binomial distribution with 2 trials. The example data provided contains the first 500 SNPs (50 samples) on chromosome 22 with a minor allele frequency of less than 2% from the European populations of the the 1000 Genomes project.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## a) Draw cuasal SNPs from a simulated genotype matrix
<span class="co"># simulate 10,000 bi-allelic SNP genotypes for 100 samples with randomly drawn </span>
<span class="co"># allele frequencies of 0.05, 0.1, 0.3 and 0.4. </span>
genotypes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/simulateGenotypes.html">simulateGenotypes</a></span>(<span class="dt">N =</span> <span class="dv">100</span>, <span class="dt">NrSNP =</span> <span class="dv">10000</span>, 
                               <span class="dt">frequencies =</span> <span class="kw">c</span>(<span class="fl">0.05</span>, <span class="fl">0.1</span>, <span class="fl">0.3</span>,<span class="fl">0.4</span>), 
                               <span class="dt">verbose =</span> <span class="ot">FALSE</span>)

<span class="co"># draw 10 causal SNPs from the genotype matrix (use non-standardised allele </span>
<span class="co"># codes i.e. (0,1,2))</span>
causalSNPs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getCausalSNPs.html">getCausalSNPs</a></span>(<span class="dt">N=</span><span class="dv">100</span>, <span class="dt">NrCausalSNPs =</span> <span class="dv">10</span>, 
                            <span class="dt">genotypes =</span> genotypes<span class="op">$</span>genotypes)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## b) Draw SNPs from external genotype files: 
<span class="co"># read genotypes from external file</span>
<span class="co"># use one of the sample genotype file provided in the </span>
<span class="co"># extdata/genotypes/subfolders (e.g.extdata/genotypes/hapgen )</span>
genotypefile &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata/genotypes/hapgen"</span>, 
                            <span class="st">"genotypes_hapgen.controls.gen"</span>, 
                            <span class="dt">package =</span> <span class="st">"PhenotypeSimulator"</span>)
<span class="co"># remove the .gen ending (oxbgen specific endings .gen and .sample are added </span>
<span class="co"># automatically )</span>
genotypefile &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">"</span><span class="ch">\\</span><span class="st">.gen"</span>,<span class="st">""</span>, genotypefile)

genotypes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/readStandardGenotypes.html">readStandardGenotypes</a></span>(<span class="dt">N =</span> <span class="dv">100</span>, <span class="dt">filename =</span> genotypefile,
                                    <span class="dt">format=</span><span class="st">"oxgen"</span>,
                                    <span class="dt">delimiter =</span> <span class="st">","</span>, <span class="dt">verbose=</span><span class="ot">TRUE</span>)

causalSNPsFromFile &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getCausalSNPs.html">getCausalSNPs</a></span>(<span class="dt">N =</span> <span class="dv">100</span>, <span class="dt">NrCausalSNPs =</span> <span class="dv">10</span>, 
                                    <span class="dt">genotypes=</span>genotypes<span class="op">$</span>genotypes)
                                    </code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## c) draw 10 causal SNPs from external genotype files: sample 10 SNPs from 
## chromosome 22
<span class="co"># use sample genotype file provided in the extdata/genotypes folder</span>
genotypeFile &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata/genotypes/"</span>, <span class="st">"genotypes_chr22.csv"</span>, 
                            <span class="dt">package =</span> <span class="st">"PhenotypeSimulator"</span>)
genoFilePrefix &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">"chr.*"</span>, <span class="st">""</span>, genotypeFile) 
genoFileSuffix &lt;-<span class="st"> ".csv"</span> 

causalSNPsSampledFromFile &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getCausalSNPs.html">getCausalSNPs</a></span>(<span class="dt">N =</span> <span class="dv">10</span>, <span class="dt">NrCausalSNPs =</span> <span class="dv">10</span>, <span class="dt">chr =</span> <span class="dv">22</span>, 
                                    <span class="dt">genoFilePrefix =</span> genoFilePrefix, 
                                    <span class="dt">genoFileSuffix =</span> genoFileSuffix,  
                                    <span class="dt">delimiter =</span> <span class="st">","</span>, <span class="dt">verbose=</span><span class="ot">TRUE</span>)
<span class="co">#&gt; Draw 10 causal SNPs from 1 chromosomes...</span>
<span class="co">#&gt; Causal chromosomes: 22</span>
<span class="co">#&gt; Get causal SNPs from chromsome-wide SNP files ( /Library/Frameworks/R.framework/Versions/3.5/Resources/library/PhenotypeSimulator/extdata/genotypes//genotypes_ ...)</span>
<span class="co">#&gt; Get 10 causal SNPs from chromsome 22 ...</span>
<span class="co">#&gt; Count number of SNPs on chromosome 22 ...</span>
<span class="co">#&gt; Sample SNPs on chromosome 22 ...</span>
<span class="co">#&gt; Sampling 10 samples from 50 genotypes provided</span></code></pre></div>
<p>The effects attached to the causal SNPs can be classified into two categories: i) a SNP can have a shared effect across traits or an independent effect across traits. The function <code>geneticFixedEffects</code> allows the user the specify what proportion <code>pIndependentGenetic</code> of SNPs should have independent effects and in the case of an independent effect, the proportion of traits <code>pTraitIndependentGenetic</code> affected by the independent effect.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># create genetic variant effects with 20% of SNPs having a specific effect, </span>
<span class="co"># affecting 40% of all simulated traits</span>
fixedGenetic &lt;-<span class="st"> </span><span class="kw"><a href="../reference/geneticFixedEffects.html">geneticFixedEffects</a></span>(<span class="dt">X_causal =</span> causalSNPsFromFile, 
                                    <span class="dt">N =</span> <span class="dv">100</span>, <span class="dt">P =</span> <span class="dv">10</span>, 
                                    <span class="dt">pIndependentGenetic =</span> <span class="fl">0.2</span>, 
                                    <span class="dt">pTraitIndependentGenetic =</span> <span class="fl">0.4</span>)</code></pre></div>
<p><img src="PhenotypeSimulator_files/figure-html/unnamed-chunk-14-1.png" width="326.4"><img src="PhenotypeSimulator_files/figure-html/unnamed-chunk-14-2.png" width="326.4"></p>
</div>
<div id="infinitesimal-genetic-effects" class="section level2">
<h2 class="hasAnchor">
<a href="#infinitesimal-genetic-effects" class="anchor"></a>2. Infinitesimal genetic effects:</h2>
<p>Infinitesimal genetic effects are simulated via <code>geneticBgEffects</code> based on the kinship estimates of the (simulated) samples. They can also be distinguished into a shared and independent effect. Both components are modeled by combining three matrix components : i) the kinship matrix <span class="math inline">\(K\)</span> [N x N] which is treated as the sample-design matrix (the genetic profile of the samples), ii) a matrix <span class="math inline">\(B\)</span> [N x P] with <span class="math inline">\(vec(B)\)</span> drawn from a normal distribution and iii) the trait design matrix <span class="math inline">\(A\)</span> [P x P]. For the independent effect, <span class="math inline">\(A\)</span> is a diagonal matrix with normally distributed values. <span class="math inline">\(A\)</span> of the shared effect is a matrix of row rank one, with normally distributed entries in row 1 and zeros elsewhere. The three matrices are multiplied to obtain the desired final effect matrix <span class="math inline">\(E\)</span>: <span class="math inline">\(E = chol(K)BA\)</span>. As for the genetic variant effects, the kinship can either be estimated from simulated genotypes or read from file. The kinship is estimated as <span class="math inline">\(K = XX_T\)</span>, with X the standardised genotypes of the samples. When estimating the kinship from the provided genotypes, the kinship is normalised by the mean of its diagonal elements and 1e-4 added to the diagonal for numerical stability. The provided kinship contains estimates for 50 samples across the entire genome.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## a) Estimate kinship from simulated genotypes
genotypes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/simulateGenotypes.html">simulateGenotypes</a></span>(<span class="dt">N =</span> <span class="dv">100</span>, <span class="dt">NrSNP =</span> <span class="dv">10000</span>, 
                               <span class="dt">frequencies =</span> <span class="kw">c</span>(<span class="fl">0.05</span>, <span class="fl">0.1</span>, <span class="fl">0.3</span>,<span class="fl">0.4</span>), 
                               <span class="dt">verbose =</span> <span class="ot">FALSE</span>)

genotypes_sd &lt;-<span class="st"> </span><span class="kw"><a href="../reference/standardiseGenotypes.html">standardiseGenotypes</a></span>(genotypes<span class="op">$</span>genotypes)
kinship &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getKinship.html">getKinship</a></span>(<span class="dt">N=</span><span class="dv">100</span>, <span class="dt">X=</span>genotypes_sd, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)

## b) Read kinship from external kinship file
kinshipFile &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata/kinship/"</span>, <span class="st">"kinship.csv"</span>, 
                           <span class="dt">package =</span> <span class="st">"PhenotypeSimulator"</span>)
kinshipFromFile &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getKinship.html">getKinship</a></span>(<span class="dt">N=</span><span class="dv">50</span>, <span class="dt">kinshipfile =</span> kinshipFile, 
                              <span class="dt">verbose =</span> <span class="ot">FALSE</span>)

genBg &lt;-<span class="st"> </span><span class="kw"><a href="../reference/geneticBgEffects.html">geneticBgEffects</a></span>(<span class="dt">N=</span><span class="dv">100</span>, <span class="dt">kinship =</span> kinship, <span class="dt">P =</span> <span class="dv">15</span>)</code></pre></div>
<p><img src="PhenotypeSimulator_files/figure-html/unnamed-chunk-16-1.png" width="326.4"><img src="PhenotypeSimulator_files/figure-html/unnamed-chunk-16-2.png" width="326.4"></p>
</div>
<div id="non-genetic-covariate-effects" class="section level2">
<h2 class="hasAnchor">
<a href="#non-genetic-covariate-effects" class="anchor"></a>3. Non-genetic covariate effects:</h2>
<p>Non-genetic covariate effects can be understood as confounding variables/covariates in an analysis, such as sex (binomial), age (normal/uniform), weight (normal) or disease status (categorical). Confounders can have effects across all traits (shared) or to a number of traits only (independent); the proportion of independent confounders from the total of simulated confounders can be chosen via <code>pIndependentConfounders</code>. The number of traits that are associated with the independent effects can be chosen via <code>pTraitIndependentConfounders</code>. Confounders can be simulated as categorical variables or following a binomial, uniform or normal distribution (as specified in <code>distConfounders</code>). Effect sizescan be simulated from a uniform or normal distribution, specified in <code>distBeta</code>. Multiple confounder sets drawn from different distributions/different parameters of the same distribution can be simulated by specifying <code>NrFixedEffects</code> and supplying the respective distribution parameters: i) <code>mConfounders</code> and <code>sdConfounders</code>: for the normal and uniform distributions, <code>mConfounders</code> is the mean/midpoint and <code>sdConfounders</code> the standard deviation/distance from midpoint, respectively; ii) <code>catConfounders</code> is the number of categorical variables to simulate; iii) <code>probConfounders</code> is the probability of success in the binomial distribution (with one trial) iv) <code>mBeta</code> and <code>sdBeta</code> are the mean/midpoint and standard deviation/distance from midpoint of the normally/uniformly distributed effect sizes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># create 1 non-genetic covariate effect affecting 30% of all simulated traits. The effect </span>
<span class="co"># follows a uniform distribution between 30 and 40  (resembling for instance age </span>
<span class="co"># in a study cohort).</span>
fixedNoiseUniform &lt;-<span class="st"> </span><span class="kw"><a href="../reference/noiseFixedEffects.html">noiseFixedEffects</a></span>(<span class="dt">N =</span> <span class="dv">100</span>, <span class="dt">P =</span> <span class="dv">10</span>, <span class="dt">NrConfounders =</span> <span class="dv">1</span>, 
                                       <span class="dt">pIndependentConfounders =</span> <span class="dv">1</span>, 
                                       <span class="dt">pTraitIndependentConfounders =</span> <span class="fl">0.3</span>, 
                                       <span class="dt">distConfounders =</span> <span class="st">"unif"</span>, 
                                       <span class="dt">mConfounders =</span> <span class="dv">35</span>, <span class="dt">sdConfounders =</span> <span class="dv">5</span>)

<span class="co"># create 2 non-genetic covariate effects with 1 specific confounder affecting </span>
<span class="co"># 20% of all simulated traits. The effects follow a normal distribution</span>
fixedNoiseNormal &lt;-<span class="st"> </span><span class="kw"><a href="../reference/noiseFixedEffects.html">noiseFixedEffects</a></span>(<span class="dt">N =</span> <span class="dv">100</span>, <span class="dt">P =</span> <span class="dv">10</span>, <span class="dt">NrConfounders =</span> <span class="dv">2</span>, 
                                      <span class="dt">pIndependentConfounders =</span> <span class="fl">0.5</span>, 
                                      <span class="dt">pTraitIndependentConfounders =</span> <span class="fl">0.2</span>, 
                                      <span class="dt">distConfounders =</span> <span class="st">"norm"</span>, 
                                      <span class="dt">mConfounders =</span> <span class="dv">0</span>, <span class="dt">sdConfounders =</span> <span class="dv">1</span>)

<span class="co"># create 1 non-genetic covariate effects affecting  all simulated traits. The </span>
<span class="co"># effect follows a binomial distribution with probability 0.5 (resembling for</span>
<span class="co"># instance sex in a study cohort).</span>
fixedNoiseBinomial &lt;-<span class="st"> </span><span class="kw"><a href="../reference/noiseFixedEffects.html">noiseFixedEffects</a></span>(<span class="dt">N =</span> <span class="dv">100</span>, <span class="dt">P =</span> <span class="dv">10</span>, <span class="dt">NrConfounders =</span> <span class="dv">1</span>, 
                                        <span class="dt">pIndependentConfounders =</span> <span class="dv">0</span>, 
                                        <span class="dt">distConfounders =</span> <span class="st">"bin"</span>, 
                                        <span class="dt">probConfounders =</span> <span class="fl">0.5</span>)</code></pre></div>
<p><img src="PhenotypeSimulator_files/figure-html/unnamed-chunk-18-1.png" width="384"><img src="PhenotypeSimulator_files/figure-html/unnamed-chunk-18-2.png" width="384"><img src="PhenotypeSimulator_files/figure-html/unnamed-chunk-18-3.png" width="384"><img src="PhenotypeSimulator_files/figure-html/unnamed-chunk-18-4.png" width="384"></p>
</div>
<div id="correlated-effects" class="section level2">
<h2 class="hasAnchor">
<a href="#correlated-effects" class="anchor"></a>4. Correlated effects:</h2>
<p><code>correlatedBgEffects</code> can be used to simulate phenotypes with a defined level of correlation between traits. For instance, such effects can reflect correlation structure decreasing in phenotypes with a spatial component. The correlation matrix can either be supplied by the user or be constructed within <em>PhenotypeSimulator</em>. Here, the level of correlation depends on the distance of the traits. Traits of distance <span class="math inline">\(d=1\)</span> (adjacent columns) will have correlation <span class="math inline">\(cor=pcorr^1\)</span>, traits with <span class="math inline">\(d=2\)</span> have <span class="math inline">\(cor=pcorr^2\)</span> up to traits with <span class="math inline">\(d=(P-1)\)</span> <span class="math inline">\(cor=pcorr^{(P-1)}\)</span>. The correlated effect <span class="math inline">\(correlated\)</span> is simulated as multivariate normal distributed with the described correlation structure <span class="math inline">\(C\)</span> as the covariance between the phenotypic traits: <span class="math inline">\(correlated ~ N_{NP}(0,C)\)</span>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># simulate correlated noise effect for 10 traits with top-level </span>
<span class="co"># correlation of 0.8</span>
correlatedNoise &lt;-<span class="st"> </span><span class="kw"><a href="../reference/correlatedBgEffects.html">correlatedBgEffects</a></span>(<span class="dt">N =</span> <span class="dv">100</span>, <span class="dt">P =</span> <span class="dv">10</span>, <span class="dt">pcorr =</span> <span class="fl">0.8</span> )

<span class="co"># correlation structure of the traits: strong the closer to the diagonal, </span>
<span class="co"># little correlation at the furthest distance to the diagonal </span>
furthestDistCorr &lt;-<span class="st"> </span><span class="fl">0.4</span><span class="op">^</span>(<span class="dv">10</span><span class="op">-</span><span class="dv">1</span>)
<span class="kw">pairs</span>(correlatedNoise<span class="op">$</span>correlatedBg, <span class="dt">pch =</span> <span class="st">"."</span>, 
      <span class="dt">main=</span><span class="kw">paste</span>(<span class="st">"Correlation at furthest distance to diagonal:</span><span class="ch">\n</span><span class="st">"</span>,
                 furthestDistCorr), <span class="dt">cex.main=</span><span class="fl">0.8</span>)
<span class="kw">image</span>(correlatedNoise<span class="op">$</span>correlatedBg, <span class="dt">main=</span><span class="st">"Correlated effects"</span>,  <span class="dt">axes=</span><span class="ot">FALSE</span>, 
      <span class="dt">cex.main=</span><span class="fl">0.8</span>)
<span class="kw">mtext</span>(<span class="dt">side =</span> <span class="dv">1</span>, <span class="dt">text =</span> <span class="st">"Samples"</span>, <span class="dt">line =</span> <span class="dv">1</span>)
<span class="kw">mtext</span>(<span class="dt">side =</span> <span class="dv">2</span>, <span class="dt">text =</span> <span class="st">"Traits"</span>, <span class="dt">line =</span> <span class="dv">1</span>)</code></pre></div>
<p><img src="PhenotypeSimulator_files/figure-html/unnamed-chunk-19-1.png" width="326.4"><img src="PhenotypeSimulator_files/figure-html/unnamed-chunk-19-2.png" width="326.4"></p>
</div>
<div id="observational-noise-effects" class="section level2">
<h2 class="hasAnchor">
<a href="#observational-noise-effects" class="anchor"></a>5. Observational noise effects:</h2>
<p>Observational noise effects are simulated as the sum of a shared and independent effect. The independent effect is simulated as <span class="math inline">\(vec(indpendent) ~ N(mean,sd)\)</span>. The shared random effect is simulated as the matrix product of two normal distributions A [N x 1] and B [P x 1]: <span class="math inline">\(shared=AB^T\)</span></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># simulate a noise random effect for 10 traits</span>
noiseBg &lt;-<span class="st"> </span><span class="kw"><a href="../reference/noiseBgEffects.html">noiseBgEffects</a></span>(<span class="dt">N =</span> <span class="dv">100</span>, <span class="dt">P =</span> <span class="dv">10</span>, <span class="dt">mean =</span> <span class="dv">0</span>, <span class="dt">sd =</span> <span class="dv">1</span>)</code></pre></div>
<p><img src="PhenotypeSimulator_files/figure-html/unnamed-chunk-21-1.png" width="326.4"><img src="PhenotypeSimulator_files/figure-html/unnamed-chunk-21-2.png" width="326.4"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#work-flow">Work-flow</a></li>
      <li>
<a href="#examples">Examples</a><ul class="nav nav-pills nav-stacked">
<li><a href="#example-1-creating-a-phenotype-composed-of-population-structure-and-observational-noise-effects-only-">Example 1: Creating a phenotype composed of population structure and observational noise effects only.</a></li>
      <li><a href="#example-2-creating-a-phenotype-composed-of-five-components">Example 2: Creating a phenotype composed of five components</a></li>
      </ul>
</li>
      <li><a href="#command-line-use">Command line use</a></li>
      <li>
<a href="#phenotype-component-functions">Phenotype component functions</a><ul class="nav nav-pills nav-stacked">
<li><a href="#genetic-variant-effects">1. Genetic variant effects:</a></li>
      <li><a href="#infinitesimal-genetic-effects">2. Infinitesimal genetic effects:</a></li>
      <li><a href="#non-genetic-covariate-effects">3. Non-genetic covariate effects:</a></li>
      <li><a href="#correlated-effects">4. Correlated effects:</a></li>
      <li><a href="#observational-noise-effects">5. Observational noise effects:</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Hannah Meyer.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
