<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using PhenotypeSimulator to simulate phenotypes based on an example dataset • PhenotypeSimulator</title>
<link rel="shortcut icon" type="image/x-icon" href="../favicon.ico">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.1/clipboard.min.js" integrity="sha256-hIvIxeqhGZF+VVeM55k0mJvWpQ6gTkWk3Emc+NmowYA=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using PhenotypeSimulator to simulate phenotypes based on an example dataset">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-standard navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">PhenotypeSimulator</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-file"></span>
     
    documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/PhenotypeSimulator.html">Step-by-step phenotype simulation</a>
    </li>
    <li>
      <a href="../articles/Simulation-and-LinearModel.html">Application of PhenotypeSimulator for a power analysis study</a>
    </li>
    <li>
      <a href="../articles/Sample-scripts-external-genotype-simulation.html">Genotype simulation supported by PhenotypeSimulator</a>
    </li>
    <li>
      <a href="../articles/SimulationBasedOnExampleData.html">Phenotype simulation based on an example dataset</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="../news/index.html">
    <span class="fa fa-newspaper-o"></span>
     
    news
  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-code"></span>
     
    functions
  </a>
</li>
<li>
  <a href="https://github.com/HannahVMeyer/PhenotypeSimulator">
    <span class="fa fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Using PhenotypeSimulator to simulate phenotypes based on an example dataset</h1>
                        <h4 class="author">Hannah Meyer</h4>
            
            <h4 class="date">26/02/2018</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/HannahVMeyer/PhenotypeSimulator/blob/master/vignettes/SimulationBasedOnExampleData.Rmd"><code>vignettes/SimulationBasedOnExampleData.Rmd</code></a></small>
      <div class="hidden name"><code>SimulationBasedOnExampleData.Rmd</code></div>

    </div>

    
    
<p><em>PhenotypeSimulator</em> allows for the simulation of high-dimensional phenotypes comprised from a number of different components: genetic variant and population- genetic effects, non-genetic covariates, correlated and observational noise effects. In this vignette, we show the usefulness of being able to specify different phenotype components with distinct properties. We demonstrate how different phenotype components relate directly to a real high-dimensional, image-derived data set of cardiac morphology.</p>
<p>Figure  shows the average left ventricular wall thickness of 1,185 healthy volunteers measured at more than 27,000 positions in the left ventricle of the heart <span class="citation">[1]</span> <span class="citation">[2]</span>.</p>
<div class="figure" style="text-align: center">
<img src="wallthickness.png" alt="\label{fig:wallthickness}**Left ventricular wall thickness and contributing factors.** Left ventricular wall thickness was measured at more than 27k positions in a cohort of 1,185 healthy volunteers. The average wall thickness at each position is depicted in saturated colours (the mesh depicts the right ventricle for reference). Heart wall thickness is determined by an interplay of genetic (green arrows) and non-genetic factors (grey arrows). Additional correlation between positional thickness can be observed for measures in close spatial proximity."><p class="caption">
<strong>Left ventricular wall thickness and contributing factors.</strong> Left ventricular wall thickness was measured at more than 27k positions in a cohort of 1,185 healthy volunteers. The average wall thickness at each position is depicted in saturated colours (the mesh depicts the right ventricle for reference). Heart wall thickness is determined by an interplay of genetic (green arrows) and non-genetic factors (grey arrows). Additional correlation between positional thickness can be observed for measures in close spatial proximity.
</p>
</div>
<p>These cardiac morphology measurements are likely shaped by a combination of direct genetic variant effects, population genetic effects and non-genetic covariates. In addition, other influences like image acquisition and processing can contribute to the final measurements - these effects are summarised as observational noise. Furthermore, by the nature of the measurements i.e. wall thickness of an approximately smooth ventricular wall, measurements in close spatial proximity will be strongly correlated.</p>
<p>The following sections will show the distribution of additional cohort measurements like height and age (non-genetic covariates) and the correlation of the measurements by proximity. In analogy to those, similar components will then be simulated with <em>PhenotypeSimulator</em>.</p>
<div id="cardiac-morphology-data" class="section level2">
<h2 class="hasAnchor">
<a href="#cardiac-morphology-data" class="anchor"></a>Cardiac morphology data</h2>
<p>The following section depicts data from 100 randomly selected participants of the 1,185 healthy volunteers of the original study. From the 27k heart wall thickness measurements, a subset of 1,000 measurements ie. phenotypes was chosen to depict the correlation pattern.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggplot2)
<span class="kw">library</span>(PhenotypeSimulator)

pheno &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">"./pheno_small.rds"</span>)
covs &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">"./covs.rds"</span>)

<span class="kw">dim</span>(pheno)</code></pre></div>
<pre><code>## [1] 1185 1000</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(covs)</code></pre></div>
<pre><code>## [1] 100   5</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">colnames</span>(covs)</code></pre></div>
<pre><code>## [1] "Sex"    "Age"    "Height" "Weight" "Slices"</code></pre>
<p>As example covariates of the dataset, age, height, weights, sex and the number slices in the 2D cardiac MRI were chosen (see original publication for details on phenotypes <span class="citation">[1]</span>). Figure  shows the distribution of these covariates across the 100 individuals.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">covs_melt &lt;-<span class="st"> </span>reshape2<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/reshape2/topics/melt">melt</a></span>(covs, <span class="dt">value.name=</span><span class="st">"meassure"</span>, 
                            <span class="dt">variable.name=</span><span class="st">"covariate"</span>)</code></pre></div>
<pre><code>## No id variables; using all as measure variables</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">covs_melt<span class="op">$</span>covariate &lt;-<span class="st"> </span><span class="kw">factor</span>(covs_melt<span class="op">$</span>covariate,
                              <span class="dt">levels=</span><span class="kw">c</span>(<span class="st">"Age"</span>, <span class="st">"Height"</span>, <span class="st">"Weight"</span>, <span class="st">"Sex"</span>, 
                                       <span class="st">"Slices"</span>),
                              <span class="dt">labels=</span> <span class="kw">c</span>(<span class="st">"Age [years]"</span>, <span class="st">"Height [cm]"</span>, 
                                        <span class="st">"Weight [kg]"</span>, <span class="st">"Sex"</span>, <span class="st">"Slices"</span>))
p_cont &lt;-<span class="st"> </span><span class="kw"><a href="ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(dplyr<span class="op">::</span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(covs_melt, covariate <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">"Age [years]"</span>, 
                                                           <span class="st">"Height [cm]"</span>, 
                                                           <span class="st">"Weight [kg]"</span>)), 
                   <span class="kw"><a href="ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x=</span>meassure))
p_cont &lt;-<span class="st"> </span>p_cont <span class="op">+</span><span class="st"> </span><span class="kw"><a href="ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span>(<span class="dt">binwidth=</span><span class="dv">1</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw"><a href="ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(<span class="op">~</span>covariate, <span class="dt">scales =</span> <span class="st">"free"</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw"><a href="ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() <span class="op">+</span>
<span class="st">    </span><span class="kw"><a href="ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="dt">axis.title.x=</span><span class="kw"><a href="ggplot2.tidyverse.org/reference/element.html">element_blank</a></span>(),
          <span class="dt">strip.background =</span> <span class="kw"><a href="ggplot2.tidyverse.org/reference/element.html">element_rect</a></span>(<span class="dt">fill=</span><span class="st">"white"</span>, <span class="dt">color=</span><span class="st">"black"</span>))

p_cat &lt;-<span class="st"> </span><span class="kw"><a href="ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(dplyr<span class="op">::</span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(covs_melt, covariate <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">"Sex"</span>, <span class="st">"Slices"</span>)), 
                   <span class="kw"><a href="ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x=</span>meassure))
p_cat &lt;-<span class="st"> </span>p_cat <span class="op">+</span><span class="st"> </span><span class="kw"><a href="ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span>(<span class="kw"><a href="ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x=</span>meassure)) <span class="op">+</span>
<span class="st">    </span><span class="kw"><a href="ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(<span class="op">~</span>covariate, <span class="dt">scales =</span> <span class="st">"free"</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw"><a href="ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() <span class="op">+</span>
<span class="st">    </span><span class="kw"><a href="ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="dt">axis.title.x=</span><span class="kw"><a href="ggplot2.tidyverse.org/reference/element.html">element_blank</a></span>(),
          <span class="dt">strip.background =</span> <span class="kw"><a href="ggplot2.tidyverse.org/reference/element.html">element_rect</a></span>(<span class="dt">fill=</span><span class="st">"white"</span>, <span class="dt">color=</span><span class="st">"black"</span>))

combine &lt;-<span class="st"> </span>cowplot<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/cowplot/topics/plot_grid">plot_grid</a></span>(p_cont, p_cat, <span class="dt">nrow=</span><span class="dv">2</span>) 
<span class="kw">print</span>(combine)</code></pre></div>
<div class="figure" style="text-align: center">
<img src="SimulationBasedOnExampleData_files/figure-html/distribution-1.png" alt="\label{fig:distribution_covs}**Distribution of covariates.** Histograms of the reduced cohort data (100 individuals) for sex, age, height, weight and number of 2D cardiac MRI slices." width="700"><p class="caption">
<strong>Distribution of covariates.</strong> Histograms of the reduced cohort data (100 individuals) for sex, age, height, weight and number of 2D cardiac MRI slices.
</p>
</div>
<p>The correlation of the 1,000 left-ventricular wall thickness measurements is shown in Figure . While we don’t know the exact influences of each covariate, the genetic effects or the observational noise, we can clearly see that there is spatial correlation effect, where positions in proximity (close to the diagonal) in general are more correlated than more distant positions.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cor_pheno &lt;-<span class="st"> </span><span class="kw">cor</span>(pheno)
gplots<span class="op">::</span><span class="kw">heatmap.2</span>(cor_pheno, <span class="dt">trace=</span><span class="st">"none"</span>, <span class="dt">dendrogram=</span><span class="st">"none"</span>, <span class="dt">keysize=</span><span class="dv">1</span>,
                  <span class="dt">col=</span><span class="kw">colorRampPalette</span>(<span class="kw">c</span>(<span class="st">"white"</span>, <span class="st">"#5BBCD6"</span>)),
                  <span class="dt">labRow=</span><span class="ot">FALSE</span>, <span class="dt">labCol=</span><span class="ot">FALSE</span>,
                  <span class="dt">breaks=</span><span class="kw">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="fl">0.001</span>),
                  <span class="dt">key.title=</span><span class="st">""</span>,
                  <span class="dt">key.xlab=</span><span class="st">"Pearson's correlation coefficient"</span>,
                  <span class="dt">key.ylab=</span><span class="st">""</span>,
                  <span class="dt">density.info=</span><span class="st">"none"</span>)</code></pre></div>
<div class="figure" style="text-align: center">
<img src="cor_pheno.png" alt="\label{fig:corpheno}\textbf{Correlation of left ventricular wall thickness measurements.}" width="\textwidth"><p class="caption">

</p>
</div>
</div>
<div id="simulation" class="section level2">
<h2 class="hasAnchor">
<a href="#simulation" class="anchor"></a>Simulation</h2>
<p>For the simulation of similar phenotype components as shown above, we make the assumption that the above distributions approximate a uniform (age), normal (weight, height) and binomial (sex) distribution. The slices are assumed to be uniformly drawn from four categories. In the following, we extract the relevant distribution parameters: mean and standard deviation for normal distribution, range for uniform distribution, and proportion of ones from binomial distribution. To approximate the spatial correlation, we get the 0.999 quantile of the correlation values.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">age_range &lt;-<span class="st"> </span><span class="kw">range</span>(covs<span class="op">$</span>Age)

weight_mean &lt;-<span class="st"> </span><span class="kw">mean</span>(covs<span class="op">$</span>Weight)
weight_sd &lt;-<span class="st"> </span><span class="kw">sd</span>(covs<span class="op">$</span>Weight)

height_mean &lt;-<span class="st"> </span><span class="kw">mean</span>(covs<span class="op">$</span>Height)
height_sd &lt;-<span class="st"> </span><span class="kw">sd</span>(covs<span class="op">$</span>Height)

slices_categories &lt;-<span class="st"> </span><span class="kw">length</span>(<span class="kw">unique</span>(covs<span class="op">$</span>Slices))

sex_proportion_male &lt;-<span class="st"> </span><span class="kw">length</span>(<span class="kw">which</span>(covs<span class="op">$</span>Sex<span class="op">==</span><span class="dv">1</span>))<span class="op">/</span><span class="kw">nrow</span>(covs)

correlation &lt;-<span class="st"> </span><span class="kw">quantile</span>(<span class="kw">unlist</span>(cor_pheno[<span class="kw">lower.tri</span>(cor_pheno)]), <span class="fl">0.999</span>)</code></pre></div>
<p>We use these distribution parameters to simulate non-genetic covariate effects with <em>PhenotypeSimulator::noiseFixedEffects</em>. For the simulation of the spatial correlation effect, we use the above quantile cut-off for the construction of the correlation matrix with <em>PhenotypeSimulator::correlatedBgEffects</em>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">25</span>)
covs_simulated &lt;-<span class="st"> </span><span class="kw"><a href="../reference/noiseFixedEffects.html">noiseFixedEffects</a></span>(<span class="dt">N=</span><span class="dv">100</span>, <span class="dt">P=</span><span class="dv">1000</span>, <span class="dt">NrFixedEffects =</span> <span class="dv">5</span>, 
                                    <span class="dt">NrConfounders =</span> <span class="dv">1</span>,
                  <span class="dt">distConfounders =</span> <span class="kw">c</span>(<span class="st">"unif"</span>, <span class="st">"norm"</span>, <span class="st">"norm"</span>, <span class="st">"cat_unif"</span>, <span class="st">"bin"</span>),
                  <span class="dt">mConfounders =</span> <span class="kw">c</span>(<span class="kw">mean</span>(age_range), weight_mean, height_mean),
                  <span class="dt">sdConfounders =</span> <span class="kw">c</span>(age_range[<span class="dv">2</span>] <span class="op">-</span><span class="kw">mean</span>(age_range), weight_sd, 
                                    height_sd),
                  <span class="dt">probConfounders =</span> sex_proportion_male,
                  <span class="dt">catConfounders =</span> slices_categories)

background_simulated &lt;-<span class="st"> </span><span class="kw"><a href="../reference/correlatedBgEffects.html">correlatedBgEffects</a></span>(<span class="dt">N=</span><span class="dv">100</span>, <span class="dt">P=</span><span class="dv">1000</span>, <span class="dt">pcorr=</span>correlation)</code></pre></div>
<p>Figure  and  show the simulated covariates and spatial correlation based on the parameters from the real data. While the distributions are not a perfect replication (as expected with small sample sizes and our initial assumptions about the true underlying distributions of the covariates), they will serve as good approximations for simulating similar datasets as the one we based the simulations on.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">covs_simulated_melt &lt;-<span class="st"> </span>reshape2<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/reshape2/topics/melt">melt</a></span>(covs_simulated<span class="op">$</span>cov, <span class="dt">value.name=</span><span class="st">"meassure"</span>, 
                            <span class="dt">variable.name=</span><span class="st">"covariate"</span>)</code></pre></div>
<pre><code>## No id variables; using all as measure variables</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">covs_simulated_melt<span class="op">$</span>covariate &lt;-<span class="st"> </span><span class="kw">factor</span>(covs_simulated_melt<span class="op">$</span>covariate,
                              <span class="dt">levels=</span><span class="kw">c</span>(<span class="st">"sharedConfounder1_unif1"</span>, 
                                       <span class="st">"sharedConfounder2_norm1"</span>, 
                                       <span class="st">"sharedConfounder3_norm1"</span>, 
                                       <span class="st">"sharedConfounder5_bin1"</span>, 
                                       <span class="st">"sharedConfounder4_cat_unif1"</span>),
                              <span class="dt">labels=</span> <span class="kw">c</span>(<span class="st">"Age simulated"</span>, <span class="st">"Height simulated"</span>, 
                                        <span class="st">"Weight simulated"</span>, <span class="st">"Sex simulated"</span>, 
                                        <span class="st">"Slices simulated"</span>))
p_cont &lt;-<span class="st"> </span><span class="kw"><a href="ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(dplyr<span class="op">::</span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(covs_simulated_melt, covariate <span class="op">%in%</span><span class="st"> </span>
<span class="st">                                   </span><span class="kw">c</span>(<span class="st">"Age simulated"</span>, <span class="st">"Height simulated"</span>, 
                                                           <span class="st">"Weight simulated"</span>)), 
                   <span class="kw"><a href="ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x=</span>meassure))
p_cont &lt;-<span class="st"> </span>p_cont <span class="op">+</span><span class="st"> </span><span class="kw"><a href="ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span>(<span class="dt">binwidth=</span><span class="dv">1</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw"><a href="ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(<span class="op">~</span>covariate, <span class="dt">scales =</span> <span class="st">"free"</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw"><a href="ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() <span class="op">+</span>
<span class="st">    </span><span class="kw"><a href="ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="dt">axis.title.x=</span><span class="kw"><a href="ggplot2.tidyverse.org/reference/element.html">element_blank</a></span>(),
          <span class="dt">strip.background =</span> <span class="kw"><a href="ggplot2.tidyverse.org/reference/element.html">element_rect</a></span>(<span class="dt">fill=</span><span class="st">"white"</span>, <span class="dt">color=</span><span class="st">"black"</span>))

p_cat &lt;-<span class="st"> </span><span class="kw"><a href="ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(dplyr<span class="op">::</span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(covs_simulated_melt, covariate <span class="op">%in%</span><span class="st"> </span>
<span class="st">                                  </span><span class="kw">c</span>(<span class="st">"Sex simulated"</span>, <span class="st">"Slices simulated"</span>)), 
                   <span class="kw"><a href="ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x=</span>meassure))
p_cat &lt;-<span class="st"> </span>p_cat <span class="op">+</span><span class="st"> </span><span class="kw"><a href="ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span>(<span class="kw"><a href="ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x=</span>meassure)) <span class="op">+</span>
<span class="st">    </span><span class="kw"><a href="ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(<span class="op">~</span>covariate, <span class="dt">scales =</span> <span class="st">"free"</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw"><a href="ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() <span class="op">+</span>
<span class="st">    </span><span class="kw"><a href="ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="dt">axis.title.x=</span><span class="kw"><a href="ggplot2.tidyverse.org/reference/element.html">element_blank</a></span>(),
          <span class="dt">strip.background =</span> <span class="kw"><a href="ggplot2.tidyverse.org/reference/element.html">element_rect</a></span>(<span class="dt">fill=</span><span class="st">"white"</span>, <span class="dt">color=</span><span class="st">"black"</span>))

combine &lt;-<span class="st"> </span>cowplot<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/cowplot/topics/plot_grid">plot_grid</a></span>(p_cont, p_cat, <span class="dt">nrow=</span><span class="dv">2</span>) 
<span class="kw">print</span>(combine)</code></pre></div>
<div class="figure" style="text-align: center">
<img src="SimulationBasedOnExampleData_files/figure-html/distribution%20simulated%20data-1.png" alt="\label{fig:distribution_simulated}**Distribution of simulated covariates.**" width="700"><p class="caption">
<strong>Distribution of simulated covariates.</strong>
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gplots<span class="op">::</span><span class="kw">heatmap.2</span>(<span class="kw">cor</span>(background_simulated<span class="op">$</span>correlatedBg), <span class="dt">trace=</span><span class="st">"none"</span>, 
                  <span class="dt">dendrogram=</span><span class="st">"none"</span>, 
                  <span class="dt">keysize=</span><span class="dv">1</span>, <span class="dt">col=</span><span class="kw">colorRampPalette</span>(<span class="kw">c</span>(<span class="st">"white"</span>, <span class="st">"#5BBCD6"</span>)),
                  <span class="dt">labRow=</span><span class="ot">FALSE</span>, <span class="dt">labCol=</span><span class="ot">FALSE</span>,
                  <span class="dt">breaks=</span><span class="kw">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="fl">0.001</span>),
                  <span class="dt">key.title=</span><span class="st">""</span>,
                  <span class="dt">key.xlab=</span><span class="st">"Pearson's correlation coefficient"</span>,
                  <span class="dt">key.ylab=</span><span class="st">""</span>,
                  <span class="dt">density.info=</span><span class="st">"none"</span>)</code></pre></div>
<div class="figure" style="text-align: center">
<img src="cor_simulated.png" alt="\label{fig:corsimulated}\textbf{Correlation of simulated spatial proximity.}" width="\textwidth"><p class="caption">

</p>
</div>

</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-deMarvao2014">
<p>1. Marvao A de, Dawes TJ, Shi W, Minas C, Keenan NG, Diamond T, et al. Population-based studies of myocardial hypertrophy: high resolution cardiovascular magnetic resonance atlases improve statistical power. Journal of Cardiovascular Magnetic Resonance. 2014;16: 16. doi:<a href="https://doi.org/10.1186/1532-429X-16-16">10.1186/1532-429X-16-16</a></p>
</div>
<div id="ref-Biffi2017">
<p>2. Biffi C, Marvao A de, Attard MI, Dawes TJ, Whiffin N, Bai W, et al. Three-dimensional Cardiovascular Imaging-Genetics: A Mass Univariate Framework. Bioinformatics. 2017; doi:<a href="https://doi.org/10.1093/bioinformatics/btx552">10.1093/bioinformatics/btx552</a></p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#cardiac-morphology-data">Cardiac morphology data</a></li>
      <li><a href="#simulation">Simulation</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Hannah Meyer.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
