<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Application of PhenotypeSimulator for a power analysis study • PhenotypeSimulator</title>
<link rel="shortcut icon" type="image/x-icon" href="../favicon.ico">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.1/clipboard.min.js" integrity="sha256-hIvIxeqhGZF+VVeM55k0mJvWpQ6gTkWk3Emc+NmowYA=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Application of PhenotypeSimulator for a power analysis study">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-standard navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">PhenotypeSimulator</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-file"></span>
     
    documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/PhenotypeSimulator.html">Step-by-step phenotype simulation</a>
    </li>
    <li>
      <a href="../articles/Simulation-and-LinearModel.html">Application of PhenotypeSimulator for a power analysis study</a>
    </li>
    <li>
      <a href="../articles/Sample-scripts-external-genotype-simulation.html">Genotype simulation supported by PhenotypeSimulator</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="../news/index.html">
    <span class="fa fa-newspaper-o"></span>
     
    news
  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-code"></span>
     
    functions
  </a>
</li>
<li>
  <a href="https://github.com/HannahVMeyer/PhenotypeSimulator">
    <span class="fa fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Application of PhenotypeSimulator for a power analysis study</h1>
                        <h4 class="author">Hannah Meyer</h4>
            
            <h4 class="date">2018-10-25</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/HannahVMeyer/PhenotypeSimulator/blob/master/vignettes/Simulation-and-LinearModel.Rmd"><code>vignettes/Simulation-and-LinearModel.Rmd</code></a></small>
      <div class="hidden name"><code>Simulation-and-LinearModel.Rmd</code></div>

    </div>

    
    
<p>This vignette demonstrates the usage and application of <em>PhenotypeSimulator</em>. It illustrates the workflow for generating a genetic and phenotypic dataset used for a simple model comparison. First, it shows an example of how to simulate genome-wide SNPs from a re-sampling based approach ( <a href="http://mathgen.stats.ox.ac.uk/genetics_software/hapgen/hapgen2.html">Hapgen2</a>). The genotypes can then be used with <em>PhenotypeSimulator</em> to generate multiple correlated phenotypes with defined genetic and non-genetic fixed effects as well as correlation induced by kinship structure. The simulated phenotypes, genotypes, non-genetic covariates and kinship are then used with a linear mixed model framework ( <a href="http://www.xzlab.org/software/GEMMAmanual.pdf">GEMMA</a>) to investigate the power of multivariate (all phenotypes are analysed jointly) and univariate models (each phenotype is analysed separately) for genome-wide association studies (GWAS).</p>
<p>Chunks that demonstrate the usage of external software in bash (Hapggen2 and GEMMA) and chunks that depend on the fully simulated genotypes or GWAS results are simply echoed and not evaluated. Chunks that show summary statistics and correlation plots are evaluated when building this vignette and use the summary data provided in the <code>extdata</code> folder of the package (the .Rmd version of the vignette shows the chunks saving the summary data).</p>
<div id="genotype-simulation-via-hapgen2" class="section level2">
<h2 class="hasAnchor">
<a href="#genotype-simulation-via-hapgen2" class="anchor"></a>Genotype simulation via Hapgen2</h2>
<p>Hapgen2 <span class="citation">[1]</span> is a resampling-based genotype simulation tool. As such it needs a reference data set from which to sample genotypes. In the following, the 1000Genomes data from the <a href="https://mathgen.stats.ox.ac.uk/impute/impute_v1.html#Using_IMPUTE_with_the_HapMap_Data">impute2 webpage</a> was used as the reference dataset. Hapgen2 and its user manual can be found and downloaded from <a href="http://mathgen.stats.ox.ac.uk/genetics_software/hapgen/hapgen2.html">here</a>.</p>
<p>The following code chunk simulates genotypes for 1000 control individuals (no cases) on chromosomes chr1 to chr22. Subsequently, the number of variants on chr5, chr7 and chr11 are counted. These chromosomes were arbitrarily chosen to contain the causal variants. We will use those numbers later as input for <em>PhenotypeSimulator</em> (providing the number of variants on the causal chromosomes is optional, but will speed up the simulation procedure).</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="va">hapdir=</span>~/data/hmeyer/Supplementary/CEU.0908.impute.files

<span class="kw">for</span> <span class="ex">chr</span> in <span class="kw">`</span><span class="fu">seq</span> 1 22<span class="kw">`;</span> <span class="kw">do</span>
    <span class="co"># Can't get hapgen to work witout the -dl flag (should be possible since </span>
    <span class="co">#v2.0.2) so create dummy variable, that can be passed on</span>
    <span class="va">dummyDL=</span><span class="kw">`</span><span class="fu">sed</span> -n <span class="st">'2'</span>p <span class="va">$hapdir</span>/CEU.0908.chr<span class="va">${chr}</span>.legend <span class="kw">|</span> <span class="fu">cut</span> -d <span class="st">' '</span>  -f 2<span class="kw">`</span>

    <span class="ex">hapgen2</span> -m <span class="va">$hapdir</span>/genetic_map_chr<span class="va">${chr}</span>_combined_b36.txt \
            -l <span class="va">$hapdir</span>/CEU.0908.chr<span class="va">${chr}</span>.legend \
            -h <span class="va">$hapdir</span>/CEU.0908.chr<span class="va">${chr}</span>.hap \
            -o <span class="va">$hapdir</span>/genotypes_chr<span class="va">${chr}</span>_hapgen \
            -n 1000 0 \
            -dl <span class="va">$dummyDL</span> 0 0 0 \
            -no_haps_output
<span class="kw">done</span>

<span class="va">chr5=</span><span class="kw">`</span><span class="fu">wc</span> -l <span class="va">$hapdir</span>/genotypes_chr5_hapgen.controls.gen <span class="kw">|</span><span class="fu">cut</span> -d <span class="st">" "</span> -f 1<span class="kw">`</span>
<span class="va">chr7=</span><span class="kw">`</span><span class="fu">wc</span> -l <span class="va">$hapdir</span>/genotypes_chr7_hapgen.controls.gen <span class="kw">|</span><span class="fu">cut</span> -d <span class="st">" "</span> -f 1<span class="kw">`</span>
<span class="va">chr11=</span><span class="kw">`</span><span class="fu">wc</span> -l <span class="va">$hapdir</span>/genotypes_chr11_hapgen.controls.gen <span class="kw">|</span><span class="fu">cut</span> -d <span class="st">" "</span> -f 1<span class="kw">`</span></code></pre></div>
<p>For the estimation of the kinship between the individuals we make use of <a href="https://www.cog-genomics.org/plink2">PLINK</a> <span class="citation">[2]</span>. By specifying –oxford-single-chr, we indicate that the input format is the ‘oxford’ format (.gen/.sample format, see <a href="http://www.stats.ox.ac.uk/~marchini/software/gwas/file_format.html">here</a>) and convert the data into plink format (–make-bed). We then strictly prune the genotypes ( –indep-pairwise and –extract) and save the chromosome-wide output file names into a file. This file is used for merging the chromosome-wide files into a single, genome-wide genotype file. Finally, we use this pruned, genome-wide SNP file to estimate the genetic relationship of the samples.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="bu">cd</span> <span class="va">$hapdir</span>

<span class="co"># convert to plink format and prune SNPs</span>
<span class="kw">for</span> <span class="ex">chr</span> in <span class="kw">`</span><span class="fu">seq</span> 1 22<span class="kw">`;</span> <span class="kw">do</span>
    <span class="ex">plink</span> --data genotypes_chr<span class="va">${chr}</span>_hapgen.controls \
        --oxford-single-chr <span class="va">$chr</span> \
        --make-bed \
        --out  genotypes_chr<span class="va">${chr}</span>_hapgen.controls
    
    <span class="ex">plink</span> --bfile genotypes_chr<span class="va">${chr}</span>_hapgen.controls \
        --indep-pairwise 50kb 5 1 \
        --out genotypes_chr<span class="va">${chr}</span>_hapgen.controls

    <span class="ex">plink</span> --bfile genotypes_chr<span class="va">${chr}</span>_hapgen.controls  \
        --extract genotypes_chr<span class="va">${chr}</span>_hapgen.controls.prune.in \
        --make-bed \
        --out genotypes_chr<span class="va">${chr}</span>_hapgen.controls.pruned

    
    <span class="bu">echo</span> -e <span class="st">"genotypes_chr</span><span class="va">${chr}</span><span class="st">_hapgen.controls.pruned"</span> <span class="op">&gt;&gt;</span> file_list
<span class="kw">done</span>

<span class="co"># Merge chromsome-wide files into a single, genome-wide file</span>
<span class="ex">plink</span> --merge-list file_list --make-bed --out genotypes_genome_hapgen.controls

<span class="kw">for</span> <span class="ex">chr</span> in <span class="kw">`</span><span class="fu">seq</span> 1 22<span class="kw">`;</span> <span class="kw">do</span>
    <span class="ex">plink</span> --bfile genotypes_chr<span class="va">${chr}</span>_hapgen.controls.pruned \
          --exclude genotypes_genome_hapgen.controls-merge.missnp \
          --make-bed \
          --out genotypes_chr<span class="va">${chr}</span>_hapgen.controls.nodup
    <span class="bu">echo</span> -e <span class="st">"genotypes_chr</span><span class="va">${chr}</span><span class="st">_hapgen.controls.nodup"</span> <span class="op">&gt;&gt;</span> file_list_nodup
<span class="kw">done</span>

<span class="ex">plink</span> --merge-list file_list_nodup --allow-no-sex \
      --make-bed \
      --out genotypes_genome_hapgen.controls

<span class="co"># compute kinship</span>
<span class="ex">plink</span> --bfile genotypes_genome_hapgen.controls\
        --make-rel square \
        --out genotypes_genome_hapgen.controls.grm</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># load kinship data, add small value to diagonal for numerical stability and</span>
<span class="co"># do an eigendecomposition</span>
indir &lt;-<span class="st"> "~/data/hmeyer/Supplementary/CEU.0908.impute.files"</span>

kinship &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="kw">paste</span>(indir, 
                            <span class="st">"/genotypes_genome_hapgen.controls.grm.rel"</span>,<span class="dt">sep=</span><span class="st">""</span>),
                      <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">header=</span><span class="ot">FALSE</span>)

kinship &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(kinship)
<span class="kw">diag</span>(kinship) &lt;-<span class="st"> </span><span class="kw">diag</span>(kinship) <span class="op">+</span><span class="st"> </span><span class="fl">1e-4</span>

kinship_decomposed &lt;-<span class="st"> </span><span class="kw">eigen</span>(kinship)
<span class="kw">write.table</span>(kinship, 
            <span class="kw">paste</span>(indir, <span class="st">"/genotypes_genome_hapgen.controls.grm.rel"</span>, 
                  <span class="dt">sep=</span><span class="st">""</span>),
            <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">col.names=</span><span class="ot">FALSE</span>, <span class="dt">row.names=</span><span class="ot">FALSE</span>)

<span class="kw">write.table</span>(kinship_decomposed<span class="op">$</span>values, 
            <span class="kw">paste</span>(indir, <span class="st">"/genotypes_eigenvalues"</span>, 
                  <span class="dt">sep=</span><span class="st">""</span>),
            <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">col.names=</span><span class="ot">FALSE</span>, <span class="dt">row.names=</span><span class="ot">FALSE</span>)

<span class="kw">write.table</span>(kinship_decomposed<span class="op">$</span>vectors, 
            <span class="kw">paste</span>(indir, <span class="st">"/genotypes_eigenvectors"</span>, 
                  <span class="dt">sep=</span><span class="st">""</span>),
            <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">col.names=</span><span class="ot">FALSE</span>, <span class="dt">row.names=</span><span class="ot">FALSE</span>)</code></pre></div>
</div>
<div id="phenotype-simulation-via-phenotypesimulator" class="section level2">
<h2 class="hasAnchor">
<a href="#phenotype-simulation-via-phenotypesimulator" class="anchor"></a>Phenotype simulation via PhenotypeSimulator</h2>
<p>We simulate a phenotype set consisting of three traits for 1000 samples with ten genetic variant effects shared across all traits (<code>theta=1</code>; SNPs randomly sampled from the simulated genotypes) and four non-genetic covariate effects. Additional correlation between the phenotypes is simulated, induced by i) the genetic structure of the individuals (infinitesimal genetic effect through estimated kinship), ii) a correleated non-genetic effect (i.e. correlated measurement noise) and iii) an observational noise effect. For each effect component its proportion of the total genetic variance is specified (the NrSNPsChromsome in the <code>runSimulation</code> function are the values we obtained from “chr5=<code>wc -l $hapdir/genotypes_chr5_hapgen.controls.gen |cut -d " " -f 1</code>” etc. above). The non-genetic covariates consist of four independent components, two following a binomial and two following a normal distribution. The genetic variant effect was simulated to only show shared effects across traits. The remainder of the components were simulated with 80% of their variance shared across all traits, while the rest remained independent. The total genetic variance accounted to 40% leaving 60% of variance explained by the noise terms.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">indir &lt;-<span class="st"> "/homes/hannah/data/hmeyer/Supplementary/CEU.0908.impute.files"</span>

<span class="co"># specify directory to save data; if it doesn't exist yet, create i.e. needs </span>
<span class="co"># writing permissions</span>
datadir &lt;-<span class="st"> '/homes/hannah/tmp/phenotypeSimulator'</span>
<span class="cf">if</span> (<span class="op">!</span><span class="kw">dir.exists</span>(datadir)) <span class="kw">dir.create</span>(datadir, <span class="dt">recursive=</span><span class="ot">TRUE</span>)

<span class="co"># specify filenames and parameters </span>
totalGeneticVar &lt;-<span class="st"> </span><span class="fl">0.4</span>
totalSNPeffect &lt;-<span class="st"> </span><span class="fl">0.1</span>
h2s &lt;-<span class="st"> </span>totalSNPeffect<span class="op">/</span>totalGeneticVar
kinshipfile &lt;-<span class="st"> </span><span class="kw">paste</span>(indir, <span class="st">"/genotypes_genome_hapgen.controls.grm.rel"</span>,
                                <span class="dt">sep=</span><span class="st">""</span>)

genoFilePrefix &lt;-<span class="st"> </span><span class="kw">paste</span>(indir, <span class="st">"/genotypes_"</span>, <span class="dt">sep=</span><span class="st">""</span>)
genoFileSuffix &lt;-<span class="st"> "_hapgen.controls.gen"</span>

<span class="co"># simulate phenotype with three phenotype components</span>
simulation &lt;-<span class="st"> </span><span class="kw"><a href="../reference/runSimulation.html">runSimulation</a></span>(<span class="dt">N =</span> <span class="dv">1000</span>, <span class="dt">P =</span> <span class="dv">3</span>, <span class="dt">cNrSNP=</span><span class="dv">10</span>, <span class="dt">seed=</span><span class="dv">43</span>,
                           <span class="dt">kinshipfile =</span> kinshipfile,
                           <span class="dt">format =</span> <span class="st">"oxgen"</span>,
                           <span class="dt">genoFilePrefix =</span> genoFilePrefix,
                           <span class="dt">genoFileSuffix =</span> genoFileSuffix,
                           <span class="dt">chr =</span> <span class="kw">c</span>(<span class="dv">5</span>,<span class="dv">7</span>,<span class="dv">11</span>),
                           <span class="dt">mBetaGenetic =</span> <span class="dv">0</span>, <span class="dt">sdBetaGenetic =</span> <span class="fl">0.2</span>,
                           <span class="dt">theta=</span><span class="dv">1</span>,
                           <span class="dt">NrSNPsOnChromosome=</span><span class="kw">c</span>(<span class="dv">533966</span>, <span class="dv">503129</span>, <span class="dv">428863</span>),
                           <span class="dt">genVar =</span> totalGeneticVar, <span class="dt">h2s =</span> h2s,
                           <span class="dt">phi =</span> <span class="fl">0.6</span>, <span class="dt">delta =</span> <span class="fl">0.2</span>, <span class="dt">rho=</span><span class="fl">0.2</span>,
                           <span class="dt">NrFixedEffects =</span> <span class="dv">2</span>, <span class="dt">NrConfounders =</span> <span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">2</span>),
                           <span class="dt">distConfounders =</span> <span class="kw">c</span>(<span class="st">"bin"</span>, <span class="st">"norm"</span>),
                           <span class="dt">probConfounders =</span> <span class="fl">0.2</span>,
                           <span class="dt">genoDelimiter=</span><span class="st">" "</span>,
                           <span class="dt">kinshipDelimiter=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,
                           <span class="dt">kinshipHeader=</span><span class="ot">FALSE</span>,
                           <span class="dt">verbose =</span> <span class="ot">TRUE</span> )</code></pre></div>
<p>Figure  shows heatmaps of the trait-trait correlation (Pearson correlation) of the final phenotype <span class="math inline">\(\mathbf{Y}\)</span> and its five phenotype components: genetic variant effects <span class="math inline">\(\mathbf{XB}\)</span>, non-genetic covariate effects <span class="math inline">\(\mathbf{WA}\)</span>, relatedness effects <span class="math inline">\(\mathbf{U}\)</span> (infinitesimal genetic effects), non-genetic correlated effects <span class="math inline">\(\mathbf{T}\)</span> and observational noise <span class="math inline">\(\mathbf{\Psi}\)</span>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cor_phenotype &lt;-<span class="st"> </span>reshape2<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/reshape2/topics/melt">melt</a></span>(<span class="kw">cor</span>(simulation<span class="op">$</span>phenoComponentsFinal<span class="op">$</span>Y))
cor_phenotype<span class="op">$</span>type &lt;-<span class="st"> "Y"</span> 

cor_genBg &lt;-<span class="st"> </span>reshape2<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/reshape2/topics/melt">melt</a></span>(<span class="kw">cor</span>(simulation<span class="op">$</span>phenoComponentsFinal<span class="op">$</span>Y_genBg))
cor_genBg<span class="op">$</span>type &lt;-<span class="st"> "U"</span> 

cor_genFixed &lt;-<span class="st"> </span>reshape2<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/reshape2/topics/melt">melt</a></span>(<span class="kw">cor</span>(simulation<span class="op">$</span>phenoComponentsFinal<span class="op">$</span>Y_genFixed))          
cor_genFixed<span class="op">$</span>type &lt;-<span class="st"> "XB"</span> 

cor_noiseFixed &lt;-<span class="st"> </span>reshape2<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/reshape2/topics/melt">melt</a></span>(<span class="kw">cor</span>(simulation<span class="op">$</span>phenoComponentsFinal<span class="op">$</span>Y_noiseFixed))          
cor_noiseFixed<span class="op">$</span>type &lt;-<span class="st"> "WA"</span> 

cor_noiseBg &lt;-<span class="st"> </span>reshape2<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/reshape2/topics/melt">melt</a></span>(<span class="kw">cor</span>(simulation<span class="op">$</span>phenoComponentsFinal<span class="op">$</span>Y_noiseBg))          
cor_noiseBg<span class="op">$</span>type &lt;-<span class="st"> "Psi"</span> 

cor_correlatedBg &lt;-<span class="st"> </span>reshape2<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/reshape2/topics/melt">melt</a></span>(<span class="kw">cor</span>(simulation<span class="op">$</span>phenoComponentsFinal<span class="op">$</span>Y_correlatedBg))          
cor_correlatedBg<span class="op">$</span>type &lt;-<span class="st"> "T"</span> 

cor_components &lt;-<span class="st"> </span><span class="kw">rbind</span>(cor_phenotype, cor_genBg, cor_genFixed, cor_noiseBg, 
                        cor_noiseFixed, cor_correlatedBg)
cor_components<span class="op">$</span>type &lt;-<span class="st"> </span><span class="kw">factor</span>(cor_components<span class="op">$</span>type, <span class="dt">levels=</span><span class="kw">c</span>(<span class="st">"Y"</span>, 
                                                            <span class="st">"WA"</span>,
                                                            <span class="st">"XB"</span>, 
                                                            <span class="st">"U"</span>, 
                                                            <span class="st">"T"</span>,
                                                            <span class="st">"Psi"</span>),
                                                    <span class="dt">labels=</span><span class="kw">c</span>(<span class="st">"bold(Y)"</span>, 
                                                            <span class="st">"bold(WA)"</span>,
                                                            <span class="st">"bold(XB)"</span>, 
                                                            <span class="st">"bold(U)"</span>, 
                                                            <span class="st">"bold(T)"</span>,
                                                            <span class="st">"bold(Psi)"</span>))
<span class="kw">colnames</span>(cor_components) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"TraitA"</span>, <span class="st">"TraitB"</span>, <span class="st">"Correlation"</span>, <span class="st">"type"</span>)

<span class="co"># For prettier plotting</span>
cor_components<span class="op">$</span>TraitA &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">"_"</span>, <span class="st">" "</span>, cor_components<span class="op">$</span>TraitA)
cor_components<span class="op">$</span>TraitB &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">"_"</span>, <span class="st">" "</span>, cor_components<span class="op">$</span>TraitB)

p_corr &lt;-<span class="st"> </span><span class="kw">ggplot</span>(<span class="dt">data=</span>cor_components, <span class="kw">aes</span>(<span class="dt">x=</span>TraitA, <span class="dt">y=</span>TraitB, <span class="dt">fill=</span>Correlation))  
p_corr &lt;-<span class="st"> </span>p_corr <span class="op">+</span><span class="st"> </span><span class="kw">geom_tile</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">facet_grid</span>(<span class="op">~</span><span class="st"> </span>type, <span class="dt">labeller =</span> label_parsed) <span class="op">+</span>
<span class="st">    </span><span class="kw">scale_x_discrete</span>(<span class="dt">expand =</span> <span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">0</span>)) <span class="op">+</span>
<span class="st">    </span><span class="kw">scale_y_discrete</span>(<span class="dt">expand =</span> <span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">0</span>)) <span class="op">+</span>
<span class="st">    </span><span class="kw">scale_fill_gradient2</span>(<span class="dt">low=</span><span class="st">"#F2AD00"</span>, <span class="dt">high=</span><span class="st">"#5BBCD6"</span> , <span class="dt">mid=</span><span class="st">"white"</span>,
                    <span class="dt">limits =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>,<span class="dv">1</span>), 
                    <span class="dt">breaks =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>, <span class="op">-</span><span class="fl">0.5</span>, <span class="dv">0</span>, <span class="fl">0.5</span>, <span class="dv">1</span>),
                    <span class="dt">labels =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>, <span class="op">-</span><span class="fl">0.5</span>, <span class="dv">0</span>, <span class="fl">0.5</span>, <span class="dv">1</span>),
                    <span class="dt">guide=</span><span class="kw">guide_colorbar</span>(<span class="dt">direction =</span> <span class="st">"horizontal"</span>,
                                         <span class="dt">title.position=</span><span class="st">"top"</span>,
                                         <span class="dt">title.hjust =</span><span class="fl">0.5</span>)) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">xlab</span>(<span class="st">" "</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">ylab</span>(<span class="st">" "</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">theme_bw</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">coord_fixed</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">axis.text.x =</span> <span class="kw">element_text</span>(<span class="dt">angle=</span><span class="dv">90</span>, <span class="dt">vjust=</span><span class="fl">0.5</span>),
            <span class="dt">legend.position =</span> <span class="st">"bottom"</span>,
            <span class="dt">strip.background =</span> <span class="kw">element_rect</span>(<span class="dt">fill=</span><span class="st">"white"</span>, <span class="dt">color=</span><span class="st">"white"</span>),
            <span class="dt">axis.text  =</span><span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">8</span>),
            <span class="dt">legend.title  =</span><span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">8</span>),
            <span class="dt">legend.text  =</span><span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">8</span>),
            <span class="dt">strip.text  =</span><span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">8</span>))
<span class="kw">print</span>(p_corr)</code></pre></div>
<div class="figure">
<img src="Simulation-and-LinearModel_files/figure-html/heatmaps-1.png" alt="\label{fig:heatmaps}Heatmaps of the trait-by-trait correlation (Pearson correlation) of the simulated phenotype and its five phenotype components." width="672"><p class="caption">
Heatmaps of the trait-by-trait correlation (Pearson correlation) of the simulated phenotype and its five phenotype components.
</p>
</div>
<p>The phenotypes, kinship, non-genetic covariates and the intermediate phenotype components (such as shared and independendent genetic variant effects) are saved to datadir. In addition to saving all components in .csv format, we also specify gemma as an output format. This option will create phenotype, kinship and covariate files that can directly be used as input for GEMMA.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Save phenotypes, kinship and non-genetic covariates in csv and </span>
<span class="co"># GEMMA-specific format</span>
outdirectory &lt;-<span class="st"> </span><span class="kw"><a href="../reference/savePheno.html">savePheno</a></span>(simulation, <span class="dt">directory =</span> datadir, 
                            <span class="dt">intercept_gemma=</span><span class="ot">TRUE</span>, <span class="dt">format=</span><span class="kw">c</span>(<span class="st">"csv"</span>, <span class="st">"gemma"</span>),
                            <span class="dt">saveIntermediate=</span><span class="ot">TRUE</span>)</code></pre></div>
<p>As the simulated genotype dataset was large, we chose to sample from the number of variants on the causal chromosomes and did not read the entire genotype data set into memory. As such, the genotypes themselves are not yet in GEMMA specific format. The following code chunk (obtained from the <a href="http://www.xzlab.org/software/GEMMAmanual.pdf">GEMMA user manual</a>) reformats the oxford format (.gen/.sample), to the GEMMA (or BIMBAM format):</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">for</span> <span class="ex">chr</span> in <span class="kw">`</span><span class="fu">seq</span> 1 22<span class="kw">`;</span> <span class="kw">do</span>
    <span class="fu">cat</span>  <span class="va">$hapdir</span>/genotypes_chr<span class="va">${chr}</span>_hapgen.controls.gen <span class="kw">|</span> <span class="fu">awk</span> -v s=1000 \
        <span class="st">'{ printf $2 "," $4 "," $5; </span>
<span class="st">        for(i=1; i&lt;=s; i++) printf "," $(i*3+3)*2+$(i*3+4); printf "\n"}'</span> \
        <span class="op">&gt;</span> <span class="va">$hapdir</span>/genotypes_chr<span class="va">${chr}</span>_hapgen.controls.gemma

<span class="kw">done</span></code></pre></div>
</div>
<div id="genome-wide-association-study-via-gemma" class="section level2">
<h2 class="hasAnchor">
<a href="#genome-wide-association-study-via-gemma" class="anchor"></a>Genome-wide association study via GEMMA</h2>
<p>One application of <em>PhenotypeSimulator</em> is generating phenotypes for model evaluation. As a case study, we show the use of the simulated phenotypes to evaluate the power of multivariate (all phenotypes are analysed jointly) and univariate models (each phenotype is analyses separately) for GWAS where multiple traits per individual are available.</p>
<p>The univariate and multivariate linear mixed model as implemented in <a href="http://www.xzlab.org/software/GEMMAmanual.pdf">GEMMA</a> <span class="citation">[3]</span> are briefly described below.</p>
<p><strong>Univariate linear mixed model</strong></p>
<p>In the univariate linear mixed model, the <span class="math inline">\(N \times 1\)</span> vector of a single phenotype from <span class="math inline">\(N\)</span> samples is modeled as the sum of a genetic fixed effect with the <span class="math inline">\(N \times 1\)</span> vector of genetic variant <span class="math inline">\(\mathbf{x}\)</span> and its effect size <span class="math inline">\(\beta\)</span>, <span class="math inline">\(\mathbf{W} = (\mathbf{w}_1, \dot, \mathbf{w}_c)\)</span> is an <span class="math inline">\(N \times C\)</span> matrix of <span class="math inline">\(C\)</span> covariates (fixed effects) including a column of 1s (implemented in <em>PhenotypeSimulator</em> via intercept_gemma=TRUE); <span class="math inline">\(\alpha\)</span> is a <span class="math inline">\(C\)</span>-vector of the corresponding coefficients including the intercept; <span class="math inline">\(\mathbf{u}\)</span> is an <span class="math inline">\(N\)</span>-vector of random genetic effects; <span class="math inline">\(\mathbf{e}\)</span> is an <span class="math inline">\(N\)</span>-vector of observational noise. <span class="math inline">\(\sigma_g\)</span> and <span class="math inline">\(\sigma_e\)</span> are the variance of the genetic and nosie random effects; <span class="math inline">\(\mathbf{K}\)</span> is the known <span class="math inline">\(N \times N\)</span> relatedness matrix and <span class="math inline">\(\mathbf{I}_n\)</span> is an <span class="math inline">\(N \times N\)</span> identity matrix. <span class="math inline">\(MVN_N\)</span> denotes the <span class="math inline">\(N\)</span>-dimensional multivariate normal distribution.</p>
<p><span class="math display">\[\mathbf{y} = \mathbf{x}\beta + \mathbf{W}\alpha + \mathbf{u} + \mathbf{e}\]</span> with <span class="math display">\[\mathbf{u} \sim MVN_{N}(0, \sigma_g \text{K})\]</span> and <span class="math display">\[\mathbf{e} \sim MVN_{N}(0, \sigma_e I_N)\]</span></p>
<p><strong>Multivariate linear mixed model</strong></p>
<p>The univariate linear mixed model can be extended to the multivariate linear mixed model with: <span class="math display">\[\mathbf{Y} = x\beta^T + \mathbf{WA} + \mathbf{U} + \mathbf{E}\]</span> with <span class="math display">\[\mathbf{U} \sim MN_{N\times P}(0, \mathbf{K}, \mathbf{V}_g)\]</span> and <span class="math display">\[\mathbf{E} \sim MN_{N\times P}(0, \mathbf{I}_N, \mathbf{V}_e)\]</span></p>
<p>where <span class="math inline">\(\mathbf{Y}\)</span> is the <span class="math inline">\(N \times P\)</span> vector of <span class="math inline">\(P\)</span> phenotype from <span class="math inline">\(N\)</span> samples, <span class="math inline">\(\mathbf{x}\)</span> the <span class="math inline">\(N \times 1\)</span> vector of genetic variant and its <span class="math inline">\(P \times 1\)</span> effect size vector <span class="math inline">\(\mathbf{\beta}\)</span>. <span class="math inline">\(\mathbf{W} = (\mathbf{w}_1, \dot, \mathbf{w}_c)\)</span> is an <span class="math inline">\(N \times C\)</span> matrix of <span class="math inline">\(C\)</span> covariates (fixed effects) including a column of 1s; <span class="math inline">\(A\)</span> is the <span class="math inline">\(C \times P\)</span>-matrix of the corresponding coefficients including the intercept; <span class="math inline">\(\mathbf{U}\)</span> is an <span class="math inline">\(N \times P\)</span>-matrix of random genetic effects; <span class="math inline">\(\mathbf{E}\)</span> is an <span class="math inline">\(N \times P\)</span>-matrix of observational noise. <span class="math inline">\(\mathbf{V}_g\)</span> and <span class="math inline">\(\mathbf{V}_e\)</span> are the <span class="math inline">\(P \times P\)</span> symmetric matrices of the genetic and noise variance components; <span class="math inline">\(\mathbf{K}\)</span> is the known <span class="math inline">\(N \times N\)</span> relatedness matrix and <span class="math inline">\(\mathbf{I}_n\)</span> is an <span class="math inline">\(N \times N\)</span> identity matrix. <span class="math inline">\(MN(\mu, \mathbf{V}_1, \mathbf{V}_2)\)</span> denotes a matrix normal distribution with mean = <span class="math inline">\(\mu\)</span>, the column covariance <span class="math inline">\(\mathbf{V}_1\)</span> and the row covariance <span class="math inline">\(\mathbf{V}_2\)</span>.</p>
<p>We use these models for the multi-trait analyses of the three simulated phenotypes (multivariate linear mixed model, flags: -lmm 2 and -n 1 2 3) and the independent analyses of each of the three phenotypes (univariate linear mixed model, flags: -lmm 2 and -n 1 or -n 2 or -n 3). We disable the automatic minor allele frequency filtering by setting -maf 0. The following code chunk shows the GEMMA (version 0.96) setup to run these models on a per chromosome basis</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="va">hapdir=</span>/homes//hannah/data/hmeyer/Supplementary/CEU.0908.impute.files
<span class="va">datadir=</span>/homes/hannah/tmp/phenotypeSimulator

<span class="kw">for</span> <span class="ex">chr</span> in <span class="kw">`</span><span class="fu">seq</span> 1 22<span class="kw">`;</span> <span class="kw">do</span>
    <span class="co"># multivariate linear mixed model across all three traits</span>
    <span class="ex">gemma0.96</span> \
    -g <span class="va">$hapdir</span>/genotypes_chr<span class="va">${chr}</span>_hapgen.controls.gemma \
    -p <span class="va">$datadir</span>/Ysim_gemma.txt \
    -d <span class="va">$hapdir</span>/genotypes_eigenvalues \
    -u <span class="va">$hapdir</span>/genotypes_eigenvectors \
    -c <span class="va">$datadir</span>/Covs_gemma.txt \
    -maf 0 \
    -km 1 \
    -lmm 2 \
    -n 1 2 3 \
    -outdir <span class="va">$datadir</span> \
    -o GWAS_gemma_LMM_mt_chr<span class="va">$chr</span>

    <span class="co"># univariate linear mixed model for trait 1</span>
    <span class="ex">gemma0.96</span> \
    -g <span class="va">$hapdir</span>/genotypes_chr<span class="va">${chr}</span>_hapgen.controls.gemma \
    -p <span class="va">$datadir</span>/Ysim_gemma.txt \
    -d <span class="va">$hapdir</span>/genotypes_eigenvalues \
    -u <span class="va">$hapdir</span>/genotypes_eigenvectors \
    -c <span class="va">$datadir</span>/Covs_gemma.txt \
    -maf 0 \
    -km 1 \
    -lmm 2 \
    -n 1 \
    -outdir <span class="va">$datadir</span> \
    -o GWAS_gemma_LMM_st_trait1_chr<span class="va">$chr</span>
   
    <span class="co"># univariate linear mixed model for trait 2</span>
     <span class="ex">gemma0.96</span> \
    -g <span class="va">$hapdir</span>/genotypes_chr<span class="va">${chr}</span>_hapgen.controls.gemma \
    -p <span class="va">$datadir</span>/Ysim_gemma.txt \
    -d <span class="va">$hapdir</span>/genotypes_eigenvalues \
    -u <span class="va">$hapdir</span>/genotypes_eigenvectors \
    -c <span class="va">$datadir</span>/Covs_gemma.txt \
    -maf 0 \
    -km 1 \
    -lmm 2 \
    -n 2 \
    -outdir <span class="va">$datadir</span> \
    -o GWAS_gemma_LMM_st_trait2_chr<span class="va">$chr</span>
    
    <span class="co"># univariate linear mixed model for trait 3</span>
    <span class="ex">gemma0.96</span> \
    -g <span class="va">$hapdir</span>/genotypes_chr<span class="va">${chr}</span>_hapgen.controls.gemma \
    -p <span class="va">$datadir</span>/Ysim_gemma.txt \
    -d <span class="va">$hapdir</span>/genotypes_eigenvalues \
    -u <span class="va">$hapdir</span>/genotypes_eigenvectors \
    -c <span class="va">$datadir</span>/Covs_gemma.txt \
    -maf 0 \
    -km 1 \
    -lmm 2 \
    -n 3 \
    -outdir <span class="va">$datadir</span> \
    -o GWAS_gemma_LMM_st_trait3_chr<span class="va">$chr</span>
<span class="kw">done</span></code></pre></div>
<p>Subsequently, we extract the columns of interest for the power analysis (simply based on p-values). The last column contains the p-value of th analyses, the second column contains the genetic variant identifier. We extract these columns and write the per-chromosome files into a genome-wide file.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">
<span class="va">columns_LMM_st=</span><span class="kw">`</span><span class="fu">head</span> -1 <span class="va">$datadir</span>/GWAS_gemma_LMM_st_trait3_chr22.assoc.txt<span class="kw">|</span><span class="fu">wc</span> -w<span class="kw">`</span>
<span class="va">columns_LMM_mt=</span><span class="kw">`</span><span class="fu">head</span> -1 <span class="va">$datadir</span>/GWAS_gemma_LMM_mt_chr22.assoc.txt <span class="kw">|</span>  <span class="fu">wc</span> -w<span class="kw">`</span>

<span class="kw">for</span> <span class="ex">chr</span> in <span class="kw">`</span><span class="fu">seq</span> 1 22<span class="kw">`;</span> <span class="kw">do</span>
    <span class="fu">tail</span> -n +2 <span class="va">$datadir</span>/GWAS_gemma_LMM_st_trait1_chr<span class="va">${chr}</span>.assoc.txt <span class="kw">|</span> <span class="fu">cut</span> \
            -f 2,<span class="va">$columns_LMM_st</span> <span class="op">&gt;&gt;</span> <span class="va">$datadir</span>/GWAS_gemma_LMM_st_trait1_pvalues.txt
    <span class="fu">tail</span> -n +2 <span class="va">$datadir</span>/GWAS_gemma_LMM_st_trait2_chr<span class="va">${chr}</span>.assoc.txt <span class="kw">|</span> <span class="fu">cut</span> \
            -f 2,<span class="va">$columns_LMM_st</span> <span class="op">&gt;&gt;</span> <span class="va">$datadir</span>/GWAS_gemma_LMM_st_trait2_pvalues.txt
    <span class="fu">tail</span> -n +2 <span class="va">$datadir</span>/GWAS_gemma_LMM_st_trait3_chr<span class="va">${chr}</span>.assoc.txt <span class="kw">|</span> <span class="fu">cut</span> \
            -f 2,<span class="va">$columns_LMM_st</span> <span class="op">&gt;&gt;</span> <span class="va">$datadir</span>/GWAS_gemma_LMM_st_trait3_pvalues.txt
    <span class="fu">tail</span> -n +2 <span class="va">$datadir</span>/GWAS_gemma_LMM_mt_chr<span class="va">${chr}</span>.assoc.txt <span class="kw">|</span> <span class="fu">cut</span> \
            -f 2,<span class="va">$columns_LMM_mt</span> <span class="op">&gt;&gt;</span> <span class="va">$datadir</span>/GWAS_gemma_LMM_mt_pvalues.txt
<span class="kw">done</span></code></pre></div>
<p>The quantile-quantile plots of p-values observed from the univariate and multivariate LMMs fitted to all genome-wide SNPs are depicted in Figure .</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">datadir=<span class="st">"/homes/hannah/tmp/phenotypeSimulator"</span>

causalSNPs &lt;-<span class="st"> </span><span class="kw">fread</span>(<span class="kw">paste</span>(datadir, <span class="st">"/SNP_NrSNP10.csv"</span>, <span class="dt">sep=</span><span class="st">""</span>), 
                    <span class="dt">header=</span><span class="ot">TRUE</span>, <span class="dt">sep=</span><span class="st">","</span>, <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>, 
                    <span class="dt">data.table=</span><span class="ot">FALSE</span>) 
causalSNPnames &lt;-<span class="st"> </span><span class="kw">colnames</span>(causalSNPs)

LMM_mt &lt;-<span class="st"> </span><span class="kw">fread</span>(<span class="kw">paste</span>(datadir, <span class="st">"/GWAS_gemma_LMM_mt_pvalues.txt"</span>, <span class="dt">sep=</span><span class="st">""</span>),
                <span class="dt">data.table=</span><span class="ot">FALSE</span>, <span class="dt">header=</span><span class="ot">FALSE</span>, <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)
LMM_st_trait1 &lt;-<span class="st"> </span><span class="kw">fread</span>(<span class="kw">paste</span>(datadir, <span class="st">"/GWAS_gemma_LMM_st_trait1_pvalues.txt"</span>, 
                            <span class="dt">sep=</span><span class="st">""</span>),
                <span class="dt">data.table=</span><span class="ot">FALSE</span>, <span class="dt">header=</span><span class="ot">FALSE</span>, <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)
LMM_st_trait2 &lt;-<span class="st"> </span><span class="kw">fread</span>(<span class="kw">paste</span>(datadir, <span class="st">"/GWAS_gemma_LMM_st_trait2_pvalues.txt"</span>, 
                            <span class="dt">sep=</span><span class="st">""</span>),
                <span class="dt">data.table=</span><span class="ot">FALSE</span>, <span class="dt">header=</span><span class="ot">FALSE</span>, <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)
LMM_st_trait3 &lt;-<span class="st"> </span><span class="kw">fread</span>(<span class="kw">paste</span>(datadir, <span class="st">"/GWAS_gemma_LMM_st_trait3_pvalues.txt"</span>, 
                            <span class="dt">sep=</span><span class="st">""</span>),
                <span class="dt">data.table=</span><span class="ot">FALSE</span>, <span class="dt">header=</span><span class="ot">FALSE</span>, <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)

LMM_mt &lt;-<span class="st"> </span>LMM_mt[<span class="kw">order</span>(LMM_mt[,<span class="dv">2</span>]),]
LMM_mt<span class="op">$</span>expected &lt;-<span class="st"> </span>stats<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/stats/topics/ppoints">ppoints</a></span>(<span class="kw">nrow</span>(LMM_mt))
LMM_mt<span class="op">$</span>model &lt;-<span class="st"> "mvLMM"</span>
LMM_mt<span class="op">$</span>trait &lt;-<span class="st"> "all"</span>
<span class="kw">colnames</span>(LMM_mt)[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>] &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"rsID"</span>, <span class="st">"observed"</span>)

LMM_st_trait1 &lt;-<span class="st"> </span>LMM_st_trait1[<span class="kw">order</span>(LMM_st_trait1[,<span class="dv">2</span>]),]
LMM_st_trait1<span class="op">$</span>expected &lt;-<span class="st"> </span>LMM_mt<span class="op">$</span>expected
LMM_st_trait1<span class="op">$</span>model &lt;-<span class="st"> "uvLMM"</span>
LMM_st_trait1<span class="op">$</span>trait &lt;-<span class="st"> "trait1"</span>
<span class="kw">colnames</span>(LMM_st_trait1)[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>] &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"rsID"</span>, <span class="st">"observed"</span>)

LMM_st_trait2 &lt;-<span class="st"> </span>LMM_st_trait2[<span class="kw">order</span>(LMM_st_trait2[,<span class="dv">2</span>]),]
LMM_st_trait2<span class="op">$</span>expected &lt;-<span class="st"> </span>LMM_mt<span class="op">$</span>expected
LMM_st_trait2<span class="op">$</span>model &lt;-<span class="st"> "uvLMM"</span>
LMM_st_trait2<span class="op">$</span>trait &lt;-<span class="st"> "trait2"</span>
<span class="kw">colnames</span>(LMM_st_trait2)[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>] &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"rsID"</span>, <span class="st">"observed"</span>)

LMM_st_trait3 &lt;-<span class="st"> </span>LMM_st_trait3[<span class="kw">order</span>(LMM_st_trait3[,<span class="dv">2</span>]),]
LMM_st_trait3<span class="op">$</span>expected &lt;-<span class="st"> </span>LMM_mt<span class="op">$</span>expected
LMM_st_trait3<span class="op">$</span>model &lt;-<span class="st"> "uvLMM"</span>
LMM_st_trait3<span class="op">$</span>trait &lt;-<span class="st"> "trait3"</span>
<span class="kw">colnames</span>(LMM_st_trait3)[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>] &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"rsID"</span>, <span class="st">"observed"</span>)

pvalues &lt;-<span class="st"> </span><span class="kw">rbind</span>(LMM_mt, LMM_st_trait1, LMM_st_trait2, LMM_st_trait3)
pvalues<span class="op">$</span>causal &lt;-<span class="st"> "all"</span>
pvalues<span class="op">$</span>causal[<span class="kw">which</span>(pvalues<span class="op">$</span>rsID <span class="op">%in%</span><span class="st"> </span>causalSNPnames)] &lt;-<span class="st"> "simulated causal"</span>
pvalues<span class="op">$</span>causal &lt;-<span class="st"> </span><span class="kw">factor</span>(pvalues<span class="op">$</span>causal, <span class="dt">levels=</span><span class="kw">c</span>(<span class="st">"all"</span>, <span class="st">"simulated causal"</span>))

p_qq &lt;-<span class="st"> </span><span class="kw">ggplot</span>(<span class="dt">data=</span>pvalues, 
                <span class="kw">aes</span>(<span class="dt">x=</span><span class="op">-</span><span class="kw">log10</span>(expected), <span class="dt">y=</span><span class="op">-</span><span class="kw">log10</span>(observed)))
p_qq &lt;-<span class="st"> </span>p_qq <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">color=</span>causal, <span class="dt">shape=</span>trait)) <span class="op">+</span>
<span class="st">    </span><span class="kw">scale_color_manual</span>(<span class="dt">values=</span><span class="kw">c</span>(<span class="st">'darkgrey'</span>, <span class="st">'#1b9e77'</span>),
                       <span class="dt">name=</span><span class="st">"SNPs"</span>,
                    <span class="dt">guide=</span><span class="kw">guide_legend</span>(<span class="dt">nrow =</span> <span class="dv">2</span>, <span class="dt">title.position=</span><span class="st">"top"</span>,
                                       <span class="dt">title.hjust =</span><span class="fl">0.5</span>,
                                       <span class="dt">override.aes =</span> <span class="kw">list</span>(<span class="dt">shape=</span><span class="dv">21</span>))) <span class="op">+</span>
<span class="st">    </span><span class="kw">scale_shape_manual</span>(<span class="dt">values=</span><span class="kw">c</span>(<span class="dv">18</span>, <span class="dv">15</span><span class="op">:</span><span class="dv">17</span>),
                       <span class="dt">name=</span><span class="st">"Traits"</span>,
                    <span class="dt">guide=</span><span class="kw">guide_legend</span>(<span class="dt">nrow =</span> <span class="dv">2</span>, 
                                       <span class="dt">override.aes =</span> <span class="kw">list</span>(<span class="dt">colour=</span><span class="st">'darkgrey'</span>),
                                       <span class="dt">title.position=</span><span class="st">"top"</span>,
                                       <span class="dt">title.hjust =</span><span class="fl">0.5</span>)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_abline</span>(<span class="dt">intercept=</span><span class="dv">0</span>, <span class="dt">slope=</span><span class="dv">1</span>, <span class="dt">col=</span><span class="st">"black"</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">data=</span>dplyr<span class="op">::</span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(pvalues, causal <span class="op">==</span><span class="st"> "simulated causal"</span>), 
                <span class="dt">color=</span><span class="st">'#1b9e77'</span>, <span class="kw">aes</span>(<span class="dt">shape=</span>trait)) <span class="op">+</span>
<span class="st">    </span><span class="kw">xlab</span>(<span class="kw">expression</span>(Expected<span class="op">~</span><span class="er">~</span><span class="op">-</span>log[<span class="dv">10</span>](<span class="kw">italic</span>(p))) ) <span class="op">+</span>
<span class="st">    </span><span class="kw">ylab</span>(<span class="kw">expression</span>(Observed<span class="op">~</span><span class="er">~</span><span class="op">-</span>log[<span class="dv">10</span>](<span class="kw">italic</span>(p)))) <span class="op">+</span>
<span class="st">    </span><span class="kw">facet_grid</span>(<span class="op">~</span><span class="st"> </span>model) <span class="op">+</span>
<span class="st">    </span><span class="kw">theme_bw</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">strip.background =</span> <span class="kw">element_rect</span>(<span class="dt">fill=</span><span class="st">"white"</span>),
            <span class="dt">axis.title =</span> <span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">12</span>), 
            <span class="dt">axis.text  =</span><span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">10</span>), 
            <span class="dt">legend.title  =</span><span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">12</span>), 
            <span class="dt">legend.text  =</span><span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">10</span>), 
            <span class="dt">strip.text  =</span><span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">10</span>),
            <span class="dt">legend.position =</span> <span class="st">"bottom"</span>,
            <span class="dt">legend.direction =</span> <span class="st">"horizontal"</span>)</code></pre></div>
<div class="figure">
<img src="../../../../../Library/Frameworks/R.framework/Versions/3.5/Resources/library/PhenotypeSimulator/extdata/resultsSimulationAndLinearModel/gwas_results_gemma_qqplot.png" alt="Quantile-quantile plots of p-values observed from the multivariate linear mixed model (mvLMM, traits:all) and the univariate linear mixed models (uvLMM, traits: trait1/trait2/trait3) fitted to each of about eight million genome-wide SNPs (grey), including the ten SNPs for which a phenotype effect was modelled (green)."><p class="caption">Quantile-quantile plots of p-values observed from the multivariate linear mixed model (mvLMM, traits:all) and the univariate linear mixed models (uvLMM, traits: trait1/trait2/trait3) fitted to each of about eight million genome-wide SNPs (grey), including the ten SNPs for which a phenotype effect was modelled (green).</p>
</div>
<p>Counting the number of causal SNPs that both models can recover, we see that the multi-trait GWAS shows greater power compared to any of the single trait analyses. Here a total of four causal SNPs (Figure ) compared the combined count of the single-trait GWAS of three was detected with a p-values of less than the commonly applied GWAS threshold of <span class="math inline">\(5 \times 10^{-8}\)</span> (Table ). The p-values of the three univariate analyses were adjusted for multiple hypothesis testing (Bonferroni adjustment).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sigSNPs &lt;-<span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(pvalues, causal <span class="op">==</span><span class="st"> "simulated causal"</span>,
                         observed <span class="op">&lt;</span><span class="st"> </span><span class="dv">5</span><span class="op">*</span><span class="dv">10</span><span class="op">^-</span><span class="dv">8</span>)
sigSNPs<span class="op">$</span>observed_adjusted &lt;-<span class="st"> </span>sigSNPs<span class="op">$</span>observed
sigSNPs<span class="op">$</span>observed_adjusted[<span class="kw">which</span>(sigSNPs<span class="op">$</span>trait <span class="op">!=</span><span class="st"> "all"</span>)] &lt;-
<span class="st">    </span><span class="dv">3</span><span class="op">*</span>sigSNPs<span class="op">$</span>observed[<span class="kw">which</span>(sigSNPs<span class="op">$</span>trait <span class="op">!=</span><span class="st"> "all"</span>)]
sigSNPs &lt;-<span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(sigSNPs, observed_adjusted <span class="op">&lt;</span><span class="st"> </span><span class="dv">5</span><span class="op">*</span><span class="dv">10</span><span class="op">^-</span><span class="dv">8</span>)</code></pre></div>
<table class="table">
<caption>P-values from multi-trait GWAS (model == mvLMM) and single-trait GWAS (model == uvLMM)</caption>
<thead><tr class="header">
<th align="left">rsID</th>
<th align="left">model</th>
<th align="left">trait</th>
<th align="right">observed_adjusted</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">rs6948782</td>
<td align="left">mvLMM</td>
<td align="left">all</td>
<td align="right">3.202576e-30</td>
</tr>
<tr class="even">
<td align="left">rs28788872</td>
<td align="left">mvLMM</td>
<td align="left">all</td>
<td align="right">8.465234e-24</td>
</tr>
<tr class="odd">
<td align="left">5-7362997</td>
<td align="left">mvLMM</td>
<td align="left">all</td>
<td align="right">3.824327e-17</td>
</tr>
<tr class="even">
<td align="left">rs10953472</td>
<td align="left">mvLMM</td>
<td align="left">all</td>
<td align="right">7.711818e-10</td>
</tr>
<tr class="odd">
<td align="left">rs6948782</td>
<td align="left">uvLMM</td>
<td align="left">trait2</td>
<td align="right">4.178292e-18</td>
</tr>
<tr class="even">
<td align="left">rs28788872</td>
<td align="left">uvLMM</td>
<td align="left">trait2</td>
<td align="right">2.721007e-14</td>
</tr>
<tr class="odd">
<td align="left">5-7362997</td>
<td align="left">uvLMM</td>
<td align="left">trait2</td>
<td align="right">6.963468e-10</td>
</tr>
</tbody>
</table>
<p>The ability of linear (mixed) models to detect the SNPs for which a phenotype effect was modelled depends on the allele frequencies of these SNPs and the effect size <span class="citation">[4]</span>,<span class="citation">[5]</span>: the higher the effect size and/or the allele frequencies the better the power to detect the SNP effects. The p-values of all SNPs with simulated effect on the phenotypes in relation to their allele frequencies and simulated effect sizes are depicted in Figure . There is a strong trend for SNPs with high allele frequencies and large simulated effect sizes to have low p-values.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Read the SNPs simulated to have an effect on the phenotype and their SNP </span>
<span class="co"># effects</span>
SNPs &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="kw">paste</span>(datadir,<span class="st">"/SNP_NrSNP10.csv"</span>, <span class="dt">sep=</span><span class="st">""</span>),
                 <span class="dt">row.names=</span><span class="dv">1</span>)
SNPeffect &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="kw">paste</span>(datadir,<span class="st">"/SNP_effects_NrSNP10.csv"</span>, <span class="dt">sep=</span><span class="st">""</span>),
                      <span class="dt">row.names=</span><span class="dv">1</span>)


<span class="co"># HapGen simulated SNPs might be encoded as 7-25481146 , which R converts to </span>
<span class="co"># X7.25481146 via make.names in read.table; substitute X and . to get original </span>
<span class="co"># names</span>
SNPnames &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">"</span><span class="ch">\\</span><span class="st">."</span>, <span class="st">"-"</span>, <span class="kw">gsub</span>(<span class="st">"X"</span>, <span class="st">""</span>, <span class="kw">colnames</span>(SNPs)))
SNPnames &lt;-<span class="st"> </span>SNPnames[<span class="kw">order</span>(SNPnames)]
simulated_withEffect &lt;-<span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(pvalues, rsID <span class="op">%in%</span><span class="st"> </span>SNPnames)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get the allele frequencies of the SNPs simulated to have an effect on the </span>
<span class="co"># phenotype</span>
freq &lt;-<span class="st"> </span><span class="kw">apply</span>(SNPs, <span class="dv">2</span>, getAlleleFrequencies)
minor &lt;-<span class="st"> </span><span class="kw">apply</span>(freq, <span class="dv">2</span>, min)
minor &lt;-<span class="st"> </span>minor[<span class="kw">order</span>(<span class="kw">gsub</span>(<span class="st">"X"</span>, <span class="st">""</span>, <span class="kw">names</span>(minor)))]

<span class="co"># format the effect size table and combine with p-value and frequency</span>
<span class="co"># information</span>
effects &lt;-<span class="st"> </span>reshape2<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/reshape2/topics/melt">melt</a></span>(SNPeffect, <span class="dt">value.name=</span><span class="st">"beta"</span>, <span class="dt">variable.name=</span><span class="st">"name"</span>)
<span class="co">#&gt; No id variables; using all as measure variables</span>
effects<span class="op">$</span>type &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">'</span><span class="ch">\\</span><span class="st">..*'</span>, <span class="st">''</span>, effects<span class="op">$</span>name)
effects<span class="op">$</span>rsID &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">"</span><span class="ch">\\</span><span class="st">."</span>, <span class="st">"-"</span>, <span class="kw">gsub</span>(<span class="st">'.*_'</span>, <span class="st">''</span>, effects<span class="op">$</span>name))
effects<span class="op">$</span>trait &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="kw">paste</span>(<span class="st">"trait"</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="dt">sep=</span><span class="st">""</span>), <span class="dv">10</span>)
effects &lt;-<span class="st"> </span>effects[<span class="kw">order</span>(effects<span class="op">$</span>rsID),]
effects &lt;-<span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(effects, rsID <span class="op">%in%</span><span class="st"> </span>simulated_withEffect<span class="op">$</span>rsID)

simulated_withEffect &lt;-<span class="st"> </span>simulated_withEffect[<span class="kw">order</span>(simulated_withEffect[, <span class="dv">1</span>]),]
simulated_withEffect<span class="op">$</span>beta &lt;-<span class="st"> </span><span class="ot">NA</span>
simulated_withEffect<span class="op">$</span>beta[simulated_withEffect<span class="op">$</span>trait <span class="op">!=</span><span class="st"> "all"</span>] &lt;-<span class="st"> </span>effects<span class="op">$</span>beta
simulated_withEffect<span class="op">$</span>freq &lt;-<span class="st"> </span><span class="kw">rep</span>(minor[<span class="kw">which</span>(SNPnames <span class="op">%in%</span><span class="st"> </span>effects<span class="op">$</span>rsID)], <span class="dt">each=</span><span class="dv">4</span>)

<span class="co"># plot association p-values in relation to the absolute value of the simulated </span>
<span class="co"># effect sizes and the SNPs allele frequencies</span>
p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(dplyr<span class="op">::</span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(simulated_withEffect, trait <span class="op">!=</span><span class="st"> "all"</span>))
p &lt;-<span class="st"> </span>p <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">x=</span>freq, <span class="dt">y=</span><span class="op">-</span><span class="kw">log10</span>(observed), <span class="dt">color=</span><span class="kw">abs</span>(beta), <span class="dt">shape=</span>trait)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_hline</span>(<span class="dt">yintercept=</span><span class="op">-</span><span class="kw">log10</span>(<span class="dv">5</span><span class="op">*</span><span class="dv">10</span><span class="op">^</span>(<span class="op">-</span><span class="dv">8</span>)), <span class="dt">color=</span><span class="st">"darkgrey"</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">scale_color_gradientn</span>(<span class="dt">colours=</span><span class="kw">c</span>(<span class="st">'#f0f9e8'</span>,<span class="st">'#ccebc5'</span>,<span class="st">'#a8ddb5'</span>,<span class="st">'#7bccc4'</span>,
                                    <span class="st">'#4eb3d3'</span>,<span class="st">'#2b8cbe'</span>,<span class="st">'#08589e'</span>),
                          <span class="dt">name=</span><span class="kw">expression</span>(<span class="st">"| simulated"</span> <span class="op">~</span>beta<span class="op">~</span><span class="st"> "|"</span>)) <span class="op">+</span>
<span class="st">    </span><span class="kw">scale_shape_discrete</span>(<span class="dt">name=</span><span class="st">"Phenotype in GWAS"</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">xlab</span>(<span class="st">"Allele frequencies"</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">ylab</span>(<span class="kw">expression</span>(<span class="op">-</span>log[<span class="dv">10</span>](p<span class="op">-</span>value))) <span class="op">+</span>
<span class="st">    </span><span class="kw">theme_bw</span>()
<span class="kw">print</span>(p)</code></pre></div>
<div class="figure">
<img src="Simulation-and-LinearModel_files/figure-html/freq-beta-1.png" alt="\label{fig:freq-beta}P-values, allele frequencies and simulated effect sizes of the ten SNPs simulated to have an effect on phenotype." width="700"><p class="caption">
P-values, allele frequencies and simulated effect sizes of the ten SNPs simulated to have an effect on phenotype.
</p>
</div>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-Su2011">
<p>1. Su Z, Marchini J, Donnelly P. HAPGEN2: simulation of multiple disease SNPs. Bioinformatics. Oxford University Press; 2011;27: 2304–2305. doi:<a href="https://doi.org/10.1093/bioinformatics/btr341">10.1093/bioinformatics/btr341</a></p>
</div>
<div id="ref-Chang2015">
<p>2. Chang CC, Chow CC, Tellier LC, Vattikuti S, Purcell SM, Lee JJ. Second-generation PLINK: rising to the challenge of larger and richer datasets. GigaScience. 2015;4: 7. doi:<a href="https://doi.org/10.1186/s13742-015-0047-8">10.1186/s13742-015-0047-8</a></p>
</div>
<div id="ref-Zhou2014">
<p>3. Zhou X, Stephens M. Efficient multivariate linear mixed model algorithms for genome-wide association studies. Nature Methods. 2014;11: 407–9. doi:<a href="https://doi.org/10.1038/nmeth.2848">10.1038/nmeth.2848</a></p>
</div>
<div id="ref-Cohen1992">
<p>4. Cohen J. A power primer. Psychological Bulletin. 1992;112: 155–9. Available: <a href="http://www.ncbi.nlm.nih.gov/pubmed/19565683" class="uri">http://www.ncbi.nlm.nih.gov/pubmed/19565683</a></p>
</div>
<div id="ref-Halsey2015">
<p>5. Halsey LG, Curran-Everett D, Vowler SL, Drummond GB. The fickle P value generates irreproducible results. Nature Methods. 2015;12: 179–185. doi:<a href="https://doi.org/10.1038/nmeth.3288">10.1038/nmeth.3288</a></p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#genotype-simulation-via-hapgen2">Genotype simulation via Hapgen2</a></li>
      <li><a href="#phenotype-simulation-via-phenotypesimulator">Phenotype simulation via PhenotypeSimulator</a></li>
      <li><a href="#genome-wide-association-study-via-gemma">Genome-wide association study via GEMMA</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Hannah Meyer.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
