<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Read genotypes from file. — readStandardGenotypes • PhenotypeSimulator</title>

<link rel="shortcut icon" type="image/x-icon" href="../favicon.ico">

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.1/clipboard.min.js" integrity="sha256-hIvIxeqhGZF+VVeM55k0mJvWpQ6gTkWk3Emc+NmowYA=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Read genotypes from file. — readStandardGenotypes" />

<meta property="og:description" content="readStandardGenotypes can read genotypes from a number of input 
formats for standard GWAS (binary plink, snptest, bimbam, gemma) or 
simulation software (binary plink, hapgen2, genome). Alternatively, simple 
text files (with specified delimiter) can be read. For more information on 
the different file formats see External genotype software and formats." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">PhenotypeSimulator</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Sample-scripts-external-genotype-simulation.html">Genotype simulation supported by *PhenotypeSimulator*</a>
    </li>
    <li>
      <a href="../articles/Simulation-and-LinearModel.html">Application of PhenotypeSimulator for a power analysis study</a>
    </li>
    <li>
      <a href="../articles/UsagePhenotypeSimulator.html">Flexible phenotype simulation with PhenotypeSimulator</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/HannahVMeyer/PhenotypeSimulator">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Read genotypes from file.</h1>
    <small class="dont-index">Source: <a href='https://github.com/HannahVMeyer/PhenotypeSimulator/blob/master/R/genotypeFunctions.R'><code>R/genotypeFunctions.R</code></a></small>
    <div class="hidden name"><code>readStandardGenotypes.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>readStandardGenotypes can read genotypes from a number of input 
formats for standard GWAS (binary plink, snptest, bimbam, gemma) or 
simulation software (binary plink, hapgen2, genome). Alternatively, simple 
text files (with specified delimiter) can be read. For more information on 
the different file formats see <em>External genotype software and formats</em>.</p>
    
    </div>

    <pre class="usage"><span class='fu'>readStandardGenotypes</span>(<span class='no'>N</span>, <span class='no'>filename</span>, <span class='kw'>format</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>sampleID</span> <span class='kw'>=</span> <span class='st'>"ID_"</span>, <span class='kw'>snpID</span> <span class='kw'>=</span> <span class='st'>"SNP_"</span>, <span class='kw'>delimiter</span> <span class='kw'>=</span> <span class='st'>","</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>N</th>
      <td><p>Number [integer] of samples to simulate.</p></td>
    </tr>
    <tr>
      <th>filename</th>
      <td><p>path/to/genotypefile [string] in plink, oxgen 
(impute2/snptest/hapgen2), genome, bimbam or [delimiter]-delimited format (
for format information see <em>External genotype software and formats</em>).</p></td>
    </tr>
    <tr>
      <th>format</th>
      <td><p>Name [string] of genotype file format. Options are: "plink", 
"oxgen", "genome", "bimbam" or "delim".</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>[boolean] If TRUE, progress info is printed to standard out.</p></td>
    </tr>
    <tr>
      <th>sampleID</th>
      <td><p>Prefix [string] for naming samples (will be followed by 
sample number from 1 to N when constructing id_samples).</p></td>
    </tr>
    <tr>
      <th>snpID</th>
      <td><p>Prefix [string] for naming SNPs (will be followed by SNP number 
from 1 to NrSNP when constructing id_snps).</p></td>
    </tr>
    <tr>
      <th>delimiter</th>
      <td><p>Field separator [string] of genotype file when 
format == "delim".</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Named list of [NrSamples X NrSNPs] genotypes (genotypes), their 
[NrSNPs] SNP IDs (id_snps), their [NrSamples] sample IDs (id_samples) and 
format-specific additional files (such as format-specific genotypes encoding 
or sample information; might be used for output writing).</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The file formats and software formates supported are described
below. For large external genotypes, consider the option to only read 
randomly selected, causal SNPs into memory via <a href='getCausalSNPs.html'>getCausalSNPs</a>.</p>
    
    <h2 class="hasAnchor" id="external-genotype-software-and-formats"><a class="anchor" href="#external-genotype-software-and-formats"></a>External genotype software and formats</h2>

    
    <h3>Formats:</h3>
    <ul>
<li><p>PLINK: consists of three files: .bed, .bim and .fam. When specifying 
the filepath, only the core of the name without the ending should be 
specified (i.e. for geno.bed, geno.bim and geno.fam, geno should be 
specified). When reading data from plink files, the absolute path to the 
plink-format file has to be provided, tilde expansion not provided.
From <a href='https://www.cog-genomics.org/plink/1.9/formats'>https://www.cog-genomics.org/plink/1.9/formats</a>:  The .bed 
files contain the  primary representation of genotype calls at biallelic 
variants in a binary format. The .bim is a text file with no header 
line, and one line  per variant with the following six fields: i) Chromosome 
code (either an integer, or 'X'/'Y'/'XY'/'MT'; '0' indicates unknown) or 
name, ii) Variant identifier, iii) Position in morgans or centimorgans (safe 
to use dummy value of '0'), iv) Base-pair coordinate (normally 1-based, but 
0 ok; limited to 231-2), v) Allele 1 (corresponding to clear bits in .bed; 
usually minor), vi) Allele 2 (corresponding to set bits in .bed; usually 
major). The .fam file is a text file with no header line, and one line per 
sample with the following six fields: i) Family ID ('FID'), ii), Within-
family ID ('IID'; cannot be '0'), iii) Within-family ID of father ('0' if 
father isn't in dataset, iv) within-family ID of mother ('0' if mother isn't 
in dataset), v) sex code ('1' = male, '2' = female, '0' = unknown), vi) 
Phenotype value ('1' = control, '2' = case, '-9'/'0'/non-numeric = missing 
data if case/control)</p></li>
<li><p>oxgen: consists of two files: the space-separated genotype file ending
in .gen and the space-separated sample file ending in .sample. When
specifying the filepath, only the core of the name without the ending should
be specified (i.e. for geno.gen and geno.sample, geno should be specified).
From
 <a href='http://www.stats.ox.ac.uk/~marchini/software/gwas/file_format.html'>http://www.stats.ox.ac.uk/~marchini/software/gwas/file_format.html</a>:
The genotype file stores data on a one-line-per-SNP format. The first five 
entries of each line should be the SNP ID, RS ID of the SNP, base-pair 
position of the SNP, the allele coded A and the allele coded B. The SNP ID 
can be used to denote the chromosome number of each SNP. The next three 
numbers on the line should be the probabilities of the three genotypes AA, 
AB and BB at the SNP for the first individual in the cohort. The next three 
numbers should be the genotype probabilities for the second individual in the 
cohort. The next three numbers are for the third individual and so on. The 
order of individuals in the genotype file should match the order of the 
individuals in the sample file. The sample file has three parts (a) a header 
line detailing the names of the columns in the file, (b) a line detailing the 
types of variables stored in each column, and (c) a line for each individual 
detailing the information for that individual. For more information on the 
sample file visit the above url or see <a href='writeStandardOutput.html'>writeStandardOutput</a>.</p></li>
<li><p>genome: The entire output of genome can be saved via `genome -options &gt;
outputfile`. The /path/to/outputfile should be provided and this function
extracts the relevant genotype information from this output file.
<a href='http://csg.sph.umich.edu/liang/genome'>http://csg.sph.umich.edu/liang/genome</a></p></li>
<li><p>bimbam: Mean genotype file format of bimbam which is a single, comma-
separated file, without information on individuals. From 
<a href='http://www.haplotype.org/bimbam.html'>http://www.haplotype.org/bimbam.html</a>: the first column of the mean 
genotype files is the SNP ID, the second and third columns are allele types 
with minor allele first. The rest columns are the mean genotypes of different 
individuals – numbers between 0 and 2 that represents the (posterior) mean 
genotype, or dosage of the minor allele.</p></li>
<li><p>delim: a [delimter]-delimited file of [NrSNPs x NrSamples] genotypes 
with the snpIDs in the first column and the sampleIDs in the first row and 
genotypes encoded as numbers between 0 and 2 representing the (posterior) 
mean genotype, or dosage of the minor allele. Can be user-genotypes or 
genotypes simulated with foward-time algorithms such as simupop 
(<a href='http://simupop.sourceforge.net/Main/HomePage'>http://simupop.sourceforge.net/Main/HomePage</a>) or MetaSim 
(<a href='project.org/web/packages/rmetasim/vignettes/CreatingLandscapes.html'>project.org/web/packages/rmetasim/vignettes/CreatingLandscapes.html</a>), 
that allow for user-specified output formats.</p></li>
</ul>
    <h3>Genotype simulation characteristics:</h3>
    <ul>
<li><p>PLINK: simple, bi-allelelic genotypes without LD structure, details can 
be found at <a href='https://www.cog-genomics.org/plink/1.9/input#simulate'>https://www.cog-genomics.org/plink/1.9/input#simulate</a>.</p></li>
<li><p>Hapgen2: resampling-based genotype simulation, details can be found at
<a href='http://mathgen.stats.ox.ac.uk/genetics_software/hapgen/hapgen2.html'>http://mathgen.stats.ox.ac.uk/genetics_software/hapgen/hapgen2.html</a>.</p></li>
<li><p>Genome: coalescent-based genotype simulation, details can be found at
<a href='http://csg.sph.umich.edu/liang/genome/GENOME-manual.pdf'>http://csg.sph.umich.edu/liang/genome/GENOME-manual.pdf</a>.</p></li>
</ul>
Examples on how to call these genotype simulation tools can be found in the 
vignette <em>sample-scripts-external-genotype-simulation</em>.
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Genome format</span>
<span class='no'>filename_genome</span>  <span class='kw'>&lt;-</span> <span class='fu'>system.file</span>(<span class='st'>"extdata/genotypes/genome/"</span>,
<span class='st'>"genotypes_genome.txt"</span>,
<span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"PhenotypeSimulator"</span>)
<span class='no'>data_genome</span> <span class='kw'>&lt;-</span> <span class='fu'>readStandardGenotypes</span>(<span class='kw'>N</span><span class='kw'>=</span><span class='fl'>100</span>, <span class='no'>filename_genome</span>, <span class='kw'>format</span> <span class='kw'>=</span><span class='st'>"genome"</span>)

<span class='no'>filename_hapgen</span>  <span class='kw'>&lt;-</span> <span class='fu'>system.file</span>(<span class='st'>"extdata/genotypes/hapgen/"</span>,
<span class='st'>"genotypes_hapgen.controls.gen"</span>,
<span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"PhenotypeSimulator"</span>)
<span class='no'>filename_hapgen</span> <span class='kw'>&lt;-</span> <span class='fu'>gsub</span>(<span class='st'>"\\.gen"</span>, <span class='st'>""</span>, <span class='no'>filename_hapgen</span>)
<span class='no'>data_hapgen</span> <span class='kw'>&lt;-</span> <span class='fu'>readStandardGenotypes</span>(<span class='kw'>N</span><span class='kw'>=</span><span class='fl'>100</span>, <span class='no'>filename_hapgen</span>, <span class='kw'>format</span><span class='kw'>=</span><span class='st'>'oxgen'</span>)

<span class='no'>filename_plink</span>  <span class='kw'>&lt;-</span> <span class='fu'>system.file</span>(<span class='st'>"extdata/genotypes/plink/"</span>,
<span class='st'>"genotypes_plink.bed"</span>,
<span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"PhenotypeSimulator"</span>)
<span class='no'>filename_plink</span> <span class='kw'>&lt;-</span> <span class='fu'>gsub</span>(<span class='st'>"\\.bed"</span>, <span class='st'>""</span>, <span class='no'>filename_plink</span>)
<span class='no'>data_plink</span> <span class='kw'>&lt;-</span> <span class='fu'>readStandardGenotypes</span>(<span class='kw'>N</span><span class='kw'>=</span><span class='fl'>100</span>, <span class='kw'>filename</span><span class='kw'>=</span><span class='no'>filename_plink</span>,
<span class='kw'>format</span><span class='kw'>=</span><span class='st'>"plink"</span>)</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#external-genotype-software-and-formats">External genotype software and formats</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Hannah Meyer.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

