<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Set simulation model. — setModel • PhenotypeSimulator</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Set simulation model. — setModel" />

<meta property="og:description" content="Based on parameters provided, this function sets the name for the phenotype 
simulation. It carries out compatibiltiy checks of the specifie parameters 
and checks for any missing information." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-standard navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">PhenotypeSimulator</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-file"></span>
     
    documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/PhenotypeSimulator.html">Step-by-step phenotype simulation</a>
    </li>
    <li>
      <a href="../articles/Simulation-and-LinearModel.html">Application of PhenotypeSimulator for a power analysis study</a>
    </li>
    <li>
      <a href="../articles/Sample-scripts-external-genotype-simulation.html">Genotype simulation supported by PhenotypeSimulator</a>
    </li>
    <li>
      <a href="../articles/SimulationBasedOnExampleData.html">Phenotype simulation based on an example dataset</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="../news/index.html">
    <span class="fa fa-newspaper-o"></span>
     
    news
  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-code"></span>
     
    functions
  </a>
</li>
<li>
  <a href="https://github.com/HannahVMeyer/PhenotypeSimulator">
    <span class="fa fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Set simulation model.</h1>
    <small class="dont-index">Source: <a href='https://github.com/HannahVMeyer/PhenotypeSimulator/blob/master/R/createphenotypeFunctions.R'><code>R/createphenotypeFunctions.R</code></a></small>
    <div class="hidden name"><code>setModel.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Based on parameters provided, this function sets the name for the phenotype 
simulation. It carries out compatibiltiy checks of the specifie parameters 
and checks for any missing information.</p>
    
    </div>

    <pre class="usage"><span class='fu'>setModel</span>(<span class='kw'>genVar</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>h2s</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>theta</span> <span class='kw'>=</span> <span class='fl'>0.8</span>, <span class='kw'>h2bg</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>eta</span> <span class='kw'>=</span> <span class='fl'>0.8</span>, <span class='kw'>noiseVar</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>delta</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>gamma</span> <span class='kw'>=</span> <span class='fl'>0.8</span>,
  <span class='kw'>rho</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>phi</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>alpha</span> <span class='kw'>=</span> <span class='fl'>0.8</span>, <span class='kw'>pcorr</span> <span class='kw'>=</span> <span class='fl'>0.6</span>,
  <span class='kw'>pIndependentConfounders</span> <span class='kw'>=</span> <span class='fl'>0.4</span>, <span class='kw'>pTraitIndependentConfounders</span> <span class='kw'>=</span> <span class='fl'>0.2</span>,
  <span class='kw'>pIndependentGenetic</span> <span class='kw'>=</span> <span class='fl'>0.4</span>, <span class='kw'>pTraitIndependentGenetic</span> <span class='kw'>=</span> <span class='fl'>0.2</span>,
  <span class='kw'>proportionNonlinear</span> <span class='kw'>=</span> <span class='fl'>0</span>, <span class='kw'>cNrSNP</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>NrConfounders</span> <span class='kw'>=</span> <span class='fl'>10</span>,
  <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>genVar</th>
      <td><p>Total genetic variance [double].</p></td>
    </tr>
    <tr>
      <th>h2s</th>
      <td><p>Proportion [double] of variance of genetic variant effects.</p></td>
    </tr>
    <tr>
      <th>theta</th>
      <td><p>Proportion [double] of variance of shared genetic variant 
effects.</p></td>
    </tr>
    <tr>
      <th>h2bg</th>
      <td><p>Proportion [double] of variance of infinitesimal genetic effects
i.e. correlation introduced by sample kinship).</p></td>
    </tr>
    <tr>
      <th>eta</th>
      <td><p>Proportion [double] of variance of shared infinitesimal genetic 
effects.</p></td>
    </tr>
    <tr>
      <th>noiseVar</th>
      <td><p>Total noise variance [double].</p></td>
    </tr>
    <tr>
      <th>delta</th>
      <td><p>Proportion [double] of variance of non-genetic covariate effect.</p></td>
    </tr>
    <tr>
      <th>gamma</th>
      <td><p>Proportion [double] of variance of shared non-genetic covariate 
effects.</p></td>
    </tr>
    <tr>
      <th>rho</th>
      <td><p>Proportion [double] of variance of correlated noise effects.</p></td>
    </tr>
    <tr>
      <th>phi</th>
      <td><p>Proportion [double] of variance of observational noise effects.</p></td>
    </tr>
    <tr>
      <th>alpha</th>
      <td><p>Proportion [double] of variance of shared observational noise 
effect.</p></td>
    </tr>
    <tr>
      <th>pcorr</th>
      <td><p>Correlation [double] between phenotypes.</p></td>
    </tr>
    <tr>
      <th>pIndependentConfounders</th>
      <td><p>Proportion [double] of non-genetic covariate 
to have a trait-independent effect.</p></td>
    </tr>
    <tr>
      <th>pTraitIndependentConfounders</th>
      <td><p>Proportion [double] of traits influenced
by independent non-genetic covariate effects.</p></td>
    </tr>
    <tr>
      <th>pIndependentGenetic</th>
      <td><p>Proportion [double] of genetic variant effects to 
have a trait-independent fixed effect.</p></td>
    </tr>
    <tr>
      <th>pTraitIndependentGenetic</th>
      <td><p>Proportion [double] of traits influenced by 
independent genetic variant effects.</p></td>
    </tr>
    <tr>
      <th>proportionNonlinear</th>
      <td><p>[double] proportion of the phenotype to be non-
linear</p></td>
    </tr>
    <tr>
      <th>cNrSNP</th>
      <td><p>Number [integer] of causal SNPs; used as genetic variant 
effects.</p></td>
    </tr>
    <tr>
      <th>NrConfounders</th>
      <td><p>Number [integer] of non-genetic covariates; used as 
non-genetic covariate effects.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>[boolean]; If TRUE, progress info is printed to standard out.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Named list containing the genetic model (modelGenetic), the noise 
model (modelNoise) and the input parameters (h2s, h2bg, noiseVar, rho, delta, 
phi, gamma, theta, eta, alpha, pcorr, proportionNonlinear). Model options 
are: modelNoise: "noNoise", "noiseFixedOnly", "noiseBgOnly", 
"noiseCorrelatedOnly", "noiseFixedAndBg","noiseCorrelatedAndBg", 
"noiseFixedAndCorrelated", "noiseFixedAndBgAndCorrelated"
modelGenetic: "noGenetic","geneticBgOnly", "geneticFixedOnly",
"geneticFixedAndBg"</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>#genetic fixed effects only</span>
<span class='no'>model</span> <span class='kw'>&lt;-</span> <span class='fu'>setModel</span>(<span class='kw'>genVar</span><span class='kw'>=</span><span class='fl'>1</span>, <span class='kw'>h2s</span><span class='kw'>=</span><span class='fl'>1</span>)</div><div class='output co'>#&gt; <span class='message'>The total noise variance (noiseVar) is: 0</span></div><div class='output co'>#&gt; <span class='message'>The noise model is: noNoise</span></div><div class='output co'>#&gt; <span class='message'>The total genetic variance (genVar) is: 1</span></div><div class='output co'>#&gt; <span class='message'>The genetic model is: geneticFixedOnly</span></div><div class='output co'>#&gt; <span class='message'>Proportion of variance of genetic variant effects (h2s): 1</span></div><div class='output co'>#&gt; <span class='message'>Proportion of variance of shared genetic variant effects (theta): 0.8</span></div><div class='output co'>#&gt; <span class='message'>Proportion of genetic variant effects to have a trait-independent fixed effect (pIndependentGenetic): 0.4</span></div><div class='output co'>#&gt; <span class='message'>Proportion of traits influenced by independent genetic variant effects (pTraitIndependentGenetic): 0.2</span></div><div class='output co'>#&gt; <span class='message'>Proportion of non-linear phenotype transformation (proportionNonlinear): 0</span></div><div class='output co'>#&gt; <span class='message'></span></div><div class='input'>
<span class='co'>#genetic fixed and bg effects</span>
<span class='no'>model</span> <span class='kw'>&lt;-</span> <span class='fu'>setModel</span>(<span class='kw'>genVar</span><span class='kw'>=</span><span class='fl'>1</span>, <span class='kw'>h2s</span><span class='kw'>=</span><span class='fl'>0.01</span>)</div><div class='output co'>#&gt; <span class='message'>The total noise variance (noiseVar) is: 0</span></div><div class='output co'>#&gt; <span class='message'>The noise model is: noNoise</span></div><div class='output co'>#&gt; <span class='message'>The total genetic variance (genVar) is: 1</span></div><div class='output co'>#&gt; <span class='message'>The genetic model is: geneticFixedAndBg</span></div><div class='output co'>#&gt; <span class='message'>Proportion of variance of genetic variant effects (h2s): 0.01</span></div><div class='output co'>#&gt; <span class='message'>Proportion of variance of shared genetic variant effects (theta): 0.8</span></div><div class='output co'>#&gt; <span class='message'>Proportion of genetic variant effects to have a trait-independent fixed effect (pIndependentGenetic): 0.4</span></div><div class='output co'>#&gt; <span class='message'>Proportion of traits influenced by independent genetic variant effects (pTraitIndependentGenetic): 0.2</span></div><div class='output co'>#&gt; <span class='message'>Proportion of variance of infinitesimal genetic effects (h2bg): 0.99</span></div><div class='output co'>#&gt; <span class='message'>Proportion of variance of shared infinitesimal genetic effects (eta): 0.8</span></div><div class='output co'>#&gt; <span class='message'>Proportion of non-linear phenotype transformation (proportionNonlinear): 0</span></div><div class='output co'>#&gt; <span class='message'></span></div><div class='input'>
<span class='co'>#genetic and noise fixed effects only</span>
<span class='no'>model</span> <span class='kw'>&lt;-</span> <span class='fu'>setModel</span>(<span class='kw'>genVar</span><span class='kw'>=</span><span class='fl'>0.4</span>, <span class='kw'>h2s</span><span class='kw'>=</span><span class='fl'>1</span>, <span class='kw'>delta</span><span class='kw'>=</span><span class='fl'>1</span>)</div><div class='output co'>#&gt; <span class='message'>The total noise variance (noiseVar) is: 0.6</span></div><div class='output co'>#&gt; <span class='message'>The noise model is: noiseFixedOnly</span></div><div class='output co'>#&gt; <span class='message'>Proportion of non-genetic covariate variance (delta): 1</span></div><div class='output co'>#&gt; <span class='message'>Proportion of variance of shared non-genetic covariate effects (gamma): 0.8</span></div><div class='output co'>#&gt; <span class='message'>Proportion of non-genetic covariates to have a trait-independent effect (pIndependentConfounders ): 0.4</span></div><div class='output co'>#&gt; <span class='message'>Proportion of traits influenced by independent non-genetic covariate effects (pTraitIndependentConfounders): 0.2</span></div><div class='output co'>#&gt; <span class='message'></span></div><div class='output co'>#&gt; <span class='message'>The total genetic variance (genVar) is: 0.4</span></div><div class='output co'>#&gt; <span class='message'>The genetic model is: geneticFixedOnly</span></div><div class='output co'>#&gt; <span class='message'>Proportion of variance of genetic variant effects (h2s): 1</span></div><div class='output co'>#&gt; <span class='message'>Proportion of variance of shared genetic variant effects (theta): 0.8</span></div><div class='output co'>#&gt; <span class='message'>Proportion of genetic variant effects to have a trait-independent fixed effect (pIndependentGenetic): 0.4</span></div><div class='output co'>#&gt; <span class='message'>Proportion of traits influenced by independent genetic variant effects (pTraitIndependentGenetic): 0.2</span></div><div class='output co'>#&gt; <span class='message'>Proportion of non-linear phenotype transformation (proportionNonlinear): 0</span></div><div class='output co'>#&gt; <span class='message'></span></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Hannah Meyer.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

